// kepler.gl/src/ai-assistant/dist/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UPDATE_AI_ASSISTANT_MESSAGES = exports.UPDATE_AI_ASSISTANT_CONFIG = exports.SET_START_SCREEN_CAPTURE = exports.SET_SCREEN_CAPTURED = exports.SET_MAP_BOUNDARY = void 0;
exports.setMapBoundary = setMapBoundary;
exports.setScreenCaptured = setScreenCaptured;
exports.setStartScreenCapture = setStartScreenCapture;
exports.updateAiAssistantConfig = updateAiAssistantConfig;
exports.updateAiAssistantMessages = updateAiAssistantMessages;
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

var ACTION_PREFIX = '@@openassistant/';
var UPDATE_AI_ASSISTANT_CONFIG = exports.UPDATE_AI_ASSISTANT_CONFIG = "".concat(ACTION_PREFIX, "UPDATE_AI_ASSISTANT_CONFIG");
var UPDATE_AI_ASSISTANT_MESSAGES = exports.UPDATE_AI_ASSISTANT_MESSAGES = "".concat(ACTION_PREFIX, "UPDATE_AI_ASSISTANT_MESSAGES");
var SET_START_SCREEN_CAPTURE = exports.SET_START_SCREEN_CAPTURE = "".concat(ACTION_PREFIX, "SET_START_SCREEN_CAPTURE");
var SET_SCREEN_CAPTURED = exports.SET_SCREEN_CAPTURED = "".concat(ACTION_PREFIX, "SET_SCREEN_CAPTURED");
var SET_MAP_BOUNDARY = exports.SET_MAP_BOUNDARY = "".concat(ACTION_PREFIX, "SET_MAP_BOUNDARY");
// Action creators
function updateAiAssistantConfig(config) {
  return {
    type: UPDATE_AI_ASSISTANT_CONFIG,
    payload: config
  };
}
function updateAiAssistantMessages(messages) {
  return {
    type: UPDATE_AI_ASSISTANT_MESSAGES,
    payload: messages
  };
}
function setStartScreenCapture(flag) {
  return {
    type: SET_START_SCREEN_CAPTURE,
    payload: flag
  };
}
function setScreenCaptured(screenshot) {
  return {
    type: SET_SCREEN_CAPTURED,
    payload: screenshot
  };
}
function setMapBoundary(nw, se) {
  return {
    type: SET_MAP_BOUNDARY,
    payload: {
      nw: nw,
      se: se
    }
  };
}
