// kepler.gl/src/ai-assistant/dist/tools/kepler-tools/basemap-tool.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BasemapToolComponent = BasemapToolComponent;
exports.basemap = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _constants = require("@kepler.gl/constants");
var _actions = require("@kepler.gl/actions");
var _utils = require("@openassistant/utils");
var _zod = require("zod");
var _react = require("react");
var _reactRedux = require("react-redux");
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

var basemap = exports.basemap = (0, _utils.extendedTool)({
  description: 'change basemap',
  parameters: _zod.z.object({
    styleType: _zod.z["enum"](['no_map', 'dark-matter', 'positron', 'voyager', 'satellite', 'dark', 'light', 'muted', 'muted_night'])
  }),
  execute: executeBasemap,
  component: BasemapToolComponent
});
function executeBasemap(_x) {
  return _executeBasemap.apply(this, arguments);
}
function _executeBasemap() {
  _executeBasemap = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(_ref) {
    var styleType;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          styleType = _ref.styleType;
          _context.prev = 1;
          if (_constants.DEFAULT_MAP_STYLES.find(function (style) {
            return style.id === styleType;
          })) {
            _context.next = 4;
            break;
          }
          throw new Error("Invalid basemap style: ".concat(styleType, "."));
        case 4:
          return _context.abrupt("return", {
            llmResult: {
              success: true,
              styleType: styleType,
              details: "basemap style changed to ".concat(styleType, ".")
            },
            additionalData: {
              styleType: styleType
            }
          });
        case 7:
          _context.prev = 7;
          _context.t0 = _context["catch"](1);
          return _context.abrupt("return", {
            llmResult: {
              success: false,
              styleType: styleType,
              details: "Error changing basemap style: ".concat(_context.t0),
              instruction: 'Try to fix the error. If the error persists, pause the execution and ask the user to try with different prompt and context.'
            }
          });
        case 10:
        case "end":
          return _context.stop();
      }
    }, _callee, null, [[1, 7]]);
  }));
  return _executeBasemap.apply(this, arguments);
}
function BasemapToolComponent(_ref2) {
  var styleType = _ref2.styleType;
  var dispatch = (0, _reactRedux.useDispatch)();
  (0, _react.useEffect)(function () {
    dispatch((0, _actions.mapStyleChange)(styleType));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  return null;
}
