// kepler.gl/src/ai-assistant/dist/tools/kepler-tools/boundary-tool.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.mapBoundary = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));
var _utils = require("@openassistant/utils");
var _zod = require("zod");
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

function isMapBoundaryContext(context) {
  return (0, _typeof2["default"])(context) === 'object' && context !== null && 'getMapBoundary' in context;
}
var mapBoundary = exports.mapBoundary = (0, _utils.extendedTool)({
  description: 'Get the boundary of the map. Northwest and Southeast coordinates in [longitude, latitude] format.',
  parameters: _zod.z.object({}),
  execute: function () {
    var _execute = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(args, options) {
      var getMapBoundary, boundary;
      return _regenerator["default"].wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            _context.prev = 0;
            if (!(!(options !== null && options !== void 0 && options.context) || !isMapBoundaryContext(options.context))) {
              _context.next = 3;
              break;
            }
            throw new Error('context getMapBoundary() not implemented.');
          case 3:
            getMapBoundary = options.context.getMapBoundary;
            boundary = getMapBoundary();
            return _context.abrupt("return", {
              llmResult: {
                success: true,
                boundary: boundary
              }
            });
          case 8:
            _context.prev = 8;
            _context.t0 = _context["catch"](0);
            return _context.abrupt("return", {
              llmResult: {
                success: false,
                error: _context.t0 instanceof Error ? _context.t0.message : 'Unknown error',
                instruction: 'Please ask the user to make sure the kepler.gl app has been intialized successfully to get the map boundary.'
              }
            });
          case 11:
          case "end":
            return _context.stop();
        }
      }, _callee, null, [[0, 8]]);
    }));
    function execute(_x, _x2) {
      return _execute.apply(this, arguments);
    }
    return execute;
  }(),
  context: {
    getMapBoundary: function getMapBoundary() {
      throw new Error('getMapBoundary() not implemented.');
    }
  }
});
