// kepler.gl/src/ai-assistant/dist/tools/kepler-tools/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getKeplerTools = getKeplerTools;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _basemapTool = require("./basemap-tool");
var _layerCreationTool = require("./layer-creation-tool");
var _layerStyleTool = require("./layer-style-tool");
var _loaddataTool = require("./loaddata-tool");
var _boundaryTool = require("./boundary-tool");
var _saveDataTool = require("./save-data-tool");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2["default"])(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; } // SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
function getKeplerTools(visState, aiAssistant) {
  // context for tools
  var getDatasets = function getDatasets() {
    return visState.datasets;
  };
  var getLayers = function getLayers() {
    return visState.layers;
  };
  var getLoaders = function getLoaders() {
    return {
      loaders: visState.loaders,
      loadOptions: visState.loadOptions
    };
  };

  // tool: addLayer
  var addLayerTool = _objectSpread(_objectSpread({}, _layerCreationTool.addLayer), {}, {
    context: {
      getDatasets: getDatasets
    }
  });

  // tool: updateLayerColor
  var updateLayerColorTool = _objectSpread(_objectSpread({}, _layerStyleTool.updateLayerColor), {}, {
    context: {
      getLayers: getLayers
    }
  });

  // tool: loadData
  var loadDataTool = _objectSpread(_objectSpread({}, _loaddataTool.loadData), {}, {
    context: {
      getLoaders: getLoaders
    },
    component: _loaddataTool.LoadDataToolComponent
  });

  // tool: mapBoundary
  var mapBoundaryTool = _objectSpread(_objectSpread({}, _boundaryTool.mapBoundary), {}, {
    context: {
      getMapBoundary: function getMapBoundary() {
        var _aiAssistant$keplerGl;
        return (_aiAssistant$keplerGl = aiAssistant.keplerGl) === null || _aiAssistant$keplerGl === void 0 ? void 0 : _aiAssistant$keplerGl.mapBoundary;
      }
    }
  });
  return {
    basemap: _basemapTool.basemap,
    addLayer: addLayerTool,
    updateLayerColor: updateLayerColorTool,
    loadData: loadDataTool,
    mapBoundary: mapBoundaryTool,
    saveDataToMap: _saveDataTool.saveToolResults
  };
}
