// kepler.gl/dist/dist/common/range-slider-timeline-panel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/node_modules/react"));
var _rangeSliderTimeline = _interopRequireDefault(require("./range-slider-timeline"));
function _getRequireWildcardCache(e) {
  if ("function" != typeof WeakMap) return null;
  var r = new WeakMap(),
    t = new WeakMap();
  return (_getRequireWildcardCache = function _getRequireWildcardCache(e) {
    return e ? t : r;
  })(e);
}
function _interopRequireWildcard(e, r) {
  if (!r && e && e.__esModule) return e;
  if (null === e || "object" != _typeof(e) && "function" != typeof e) return {
    "default": e
  };
  var t = _getRequireWildcardCache(r);
  if (t && t.has(e)) return t.get(e);
  var n = {
      __proto__: null
    },
    a = Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) {
    var i = a ? Object.getOwnPropertyDescriptor(e, u) : null;
    i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u];
  }
  return n["default"] = e, t && t.set(e, n), n;
}
function ownKeys(e, r) {
  var t = Object.keys(e);
  if (Object.getOwnPropertySymbols) {
    var o = Object.getOwnPropertySymbols(e);
    r && (o = o.filter(function (r) {
      return Object.getOwnPropertyDescriptor(e, r).enumerable;
    })), t.push.apply(t, o);
  }
  return t;
}
function _objectSpread(e) {
  for (var r = 1; r < arguments.length; r++) {
    var t = null != arguments[r] ? arguments[r] : {};
    r % 2 ? ownKeys(Object(t), !0).forEach(function (r) {
      (0, _defineProperty2["default"])(e, r, t[r]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) {
      Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r));
    });
  }
  return e;
} // SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
function RangeSliderSubAnimationPanelFactory(RangeSliderTimeline) {
  var RangeSliderSubAnimationPanel = function RangeSliderSubAnimationPanel(_ref) {
    var subAnimations = _ref.subAnimations,
      scaledValue = _ref.scaledValue,
      style = _ref.style;
    var containerStyle = (0, _react.useMemo)(function () {
      return _objectSpread({
        display: 'flex',
        justifyContent: 'spaceBetween',
        flexWrap: 'wrap'
      }, style);
    }, [style]);
    return /*#__PURE__*/_react["default"].createElement("div", {
      style: containerStyle
    }, subAnimations.map(function (subAnimation, index) {
      return /*#__PURE__*/_react["default"].createElement(RangeSliderTimeline, {
        key: index,
        subAnimation: subAnimation,
        scaledValue: scaledValue
      });
    }));
  };
  return RangeSliderSubAnimationPanel;
}
RangeSliderSubAnimationPanelFactory.deps = [_rangeSliderTimeline["default"]];
var _default = exports["default"] = RangeSliderSubAnimationPanelFactory;
