// kepler.gl/dist/dist/common/sync-timeline-control.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.SYNC_TIMELINE_ANIMATION_ITEMS = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireDefault(require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/node_modules/react"));
var _styledComponents = require("../common/styled-components");
var _localization = require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/src/localization/src");
var _constants = require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/src/constants/src");
var _iconButton = _interopRequireDefault(require("./icon-button"));
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

var SYNC_TIMELINE_ANIMATION_ITEMS = exports.SYNC_TIMELINE_ANIMATION_ITEMS = (0, _defineProperty2["default"])((0, _defineProperty2["default"])({}, _constants.SYNC_TIMELINE_MODES.start, {
  id: _constants.SYNC_TIMELINE_MODES.start,
  content: function content() {
    return /*#__PURE__*/_react["default"].createElement("span", null, "Start");
  },
  tooltip: 'tooltip.syncTimelineStart'
}), _constants.SYNC_TIMELINE_MODES.end, {
  id: _constants.SYNC_TIMELINE_MODES.end,
  content: function content() {
    return /*#__PURE__*/_react["default"].createElement("span", null, "End");
  },
  tooltip: 'tooltip.syncTimelineEnd'
});
function SyncTimelineControlFactory() {
  var SyncTimelineControl = function SyncTimelineControl(_ref) {
    var _ref$syncTimelineAnim = _ref.syncTimelineAnimationItems,
      syncTimelineAnimationItems = _ref$syncTimelineAnim === void 0 ? SYNC_TIMELINE_ANIMATION_ITEMS : _ref$syncTimelineAnim,
      syncTimelineMode = _ref.syncTimelineMode,
      btnStyle = _ref.btnStyle,
      setFilterSyncTimelineMode = _ref.setFilterSyncTimelineMode;
    return /*#__PURE__*/_react["default"].createElement("div", null, Object.values(syncTimelineAnimationItems).filter(function (item, _index) {
      return item.id !== syncTimelineMode;
    }).map(function (item) {
      return /*#__PURE__*/_react["default"].createElement(_iconButton["default"], (0, _extends2["default"])({
        key: item.id,
        "data-tip": true,
        "data-for": "".concat(item.id, "-tooltip"),
        className: "playback-control-button",
        onClick: function onClick() {
          return setFilterSyncTimelineMode(item.id);
        }
      }, btnStyle), /*#__PURE__*/_react["default"].createElement(item.content, null), item.tooltip ? /*#__PURE__*/_react["default"].createElement(_styledComponents.Tooltip, {
        id: "".concat(item.id, "-tooltip"),
        effect: "solid",
        place: "top"
      }, /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
        id: item.tooltip
      })) : null);
    }));
  };
  return SyncTimelineControl;
}
var _default = exports["default"] = SyncTimelineControlFactory;
