// kepler.gl/dist/dist/modals/cloud-tile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.StyledWarning = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));
var _react = _interopRequireWildcard(require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/node_modules/react"));
var _styledComponents = _interopRequireDefault(require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/node_modules/styled-components"));
var _icons = require("../common/icons");
var _styledComponents2 = require("../common/styled-components");
var _useCloudListProvider2 = require("../hooks/use-cloud-list-provider");
var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6; // SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
function _getRequireWildcardCache(e) {
  if ("function" != typeof WeakMap) return null;
  var r = new WeakMap(),
    t = new WeakMap();
  return (_getRequireWildcardCache = function _getRequireWildcardCache(e) {
    return e ? t : r;
  })(e);
}
function _interopRequireWildcard(e, r) {
  if (!r && e && e.__esModule) return e;
  if (null === e || "object" != _typeof(e) && "function" != typeof e) return {
    "default": e
  };
  var t = _getRequireWildcardCache(r);
  if (t && t.has(e)) return t.get(e);
  var n = {
      __proto__: null
    },
    a = Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) {
    var i = a ? Object.getOwnPropertyDescriptor(e, u) : null;
    i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u];
  }
  return n["default"] = e, t && t.set(e, n), n;
}
var StyledTileWrapper = _styledComponents["default"].div.attrs({
  className: 'provider-tile__wrapper'
})(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  flex-direction: column;\n  align-items: center;\n  justify-content: flex-start;\n  border-radius: 2px;\n  border: 1px solid\n    ", ";\n  color: ", ";\n  cursor: pointer;\n  font-weight: 500;\n  width: 120px;\n  height: 168px;\n  background-color: #ffffff;\n  transition: ", ";\n  position: relative;\n  &:hover {\n    border: 1px solid ", ";\n    color: ", ";\n  }\n\n  .button {\n    margin-top: 20px;\n  }\n"])), function (props) {
  return props.selected ? props.theme.primaryBtnBgd : props.theme.selectBorderColorLT;
}, function (props) {
  return props.selected ? props.theme.primaryBtnBgd : props.theme.selectBorderColorLT;
}, function (props) {
  return props.theme.transition;
}, function (props) {
  return props.theme.primaryBtnBgd;
}, function (props) {
  return props.theme.primaryBtnBgd;
});
var StyledBox = (0, _styledComponents["default"])(_styledComponents2.CenterVerticalFlexbox)(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  margin-right: 12px;\n  position: relative;\n"])));
var StyledCloudName = _styledComponents["default"].div(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2["default"])(["\n  font-size: 12px;\n  margin-top: 12px;\n  margin-bottom: 4px;\n"])));
var StyledUserName = _styledComponents["default"].div(_templateObject4 || (_templateObject4 = (0, _taggedTemplateLiteral2["default"])(["\n  font-size: 11px;\n  margin-top: 8px;\n  text-align: center;\n  color: ", ";\n  overflow: hidden;\n  width: 100px;\n  text-overflow: ellipsis;\n"])), function (props) {
  return props.theme.primaryBtnActBgd;
});
var LoginButton = function LoginButton(_ref) {
  var onClick = _ref.onClick;
  return /*#__PURE__*/_react["default"].createElement(_styledComponents2.Button, {
    className: "login-button",
    link: true,
    small: true,
    onClick: onClick
  }, /*#__PURE__*/_react["default"].createElement(_icons.Login, null), "Login");
};
var LogoutButton = function LogoutButton(_ref2) {
  var onClick = _ref2.onClick;
  return /*#__PURE__*/_react["default"].createElement(_styledComponents2.Button, {
    className: "logout-button",
    link: true,
    small: true,
    onClick: onClick
  }, /*#__PURE__*/_react["default"].createElement(_icons.Logout, null), "Logout");
};
var NewTag = _styledComponents["default"].div(_templateObject5 || (_templateObject5 = (0, _taggedTemplateLiteral2["default"])(["\n  width: 37px;\n  height: 19px;\n  display: flex;\n  align-content: center;\n  justify-content: center;\n  border-radius: 8px;\n  padding: 4px 8px;\n  background-color: #EDE9F9;\n  color: #8863F8;\n  position: absolute;\n  left: 35%;\n  top: -8px\n  z-index: 500;\n  font-size: 11px;\n  line-height: 10px;\n"])));
var StyledWarning = exports.StyledWarning = _styledComponents["default"].span(_templateObject6 || (_templateObject6 = (0, _taggedTemplateLiteral2["default"])(["\n  color: ", ";\n  font-weight: ", ";\n"])), function (props) {
  return props.theme.errorColor;
}, function (props) {
  return props.theme.selectFontWeightBold;
});
/**
 * this component display a provider and allows users to select and set the current provider
 * @param provider
 * @param actionName
 * @constructor
 */
var CloudTile = function CloudTile(_ref3) {
  var provider = _ref3.provider,
    actionName = _ref3.actionName;
  var _useState = (0, _react.useState)(null),
    _useState2 = (0, _slicedToArray2["default"])(_useState, 2),
    user = _useState2[0],
    setUser = _useState2[1];
  var _useState3 = (0, _react.useState)(null),
    _useState4 = (0, _slicedToArray2["default"])(_useState3, 2),
    error = _useState4[0],
    setError = _useState4[1];
  var _useState5 = (0, _react.useState)(false),
    _useState6 = (0, _slicedToArray2["default"])(_useState5, 2),
    isLoading = _useState6[0],
    setIsLoading = _useState6[1];
  var _useCloudListProvider = (0, _useCloudListProvider2.useCloudListProvider)(),
    currentProvider = _useCloudListProvider.provider,
    setProvider = _useCloudListProvider.setProvider;
  var isSelected = provider === currentProvider;
  (0, _react.useEffect)(function () {
    if (!provider) {
      return;
    }
    setError(null);
    setIsLoading(true);
    setError(null);
    provider.getUser().then(setUser)["catch"](setError)["finally"](function () {
      setError(null);
      setIsLoading(false);
    });
  }, [provider]);
  var onLogin = (0, _react.useCallback)( /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
    var _user;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          setError(null);
          setIsLoading(true);
          _context.prev = 2;
          _context.next = 5;
          return provider.login();
        case 5:
          _user = _context.sent;
          setUser(_user);
          setProvider(provider);
          _context.next = 13;
          break;
        case 10:
          _context.prev = 10;
          _context.t0 = _context["catch"](2);
          setError(_context.t0);
        case 13:
          setIsLoading(false);
        case 14:
        case "end":
          return _context.stop();
      }
    }, _callee, null, [[2, 10]]);
  })), [provider, setProvider]);
  var onSelect = (0, _react.useCallback)( /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
    return _regenerator["default"].wrap(function _callee2$(_context2) {
      while (1) switch (_context2.prev = _context2.next) {
        case 0:
          if (!isLoading) {
            _context2.next = 2;
            break;
          }
          return _context2.abrupt("return");
        case 2:
          if (!user) {
            _context2.next = 5;
            break;
          }
          setProvider(provider);
          return _context2.abrupt("return");
        case 5:
          _context2.prev = 5;
          _context2.next = 8;
          return onLogin();
        case 8:
          setProvider(provider);
          _context2.next = 15;
          break;
        case 11:
          _context2.prev = 11;
          _context2.t0 = _context2["catch"](5);
          setError(_context2.t0);
          setProvider(null);
        case 15:
        case "end":
          return _context2.stop();
      }
    }, _callee2, null, [[5, 11]]);
  })), [setProvider, provider, user, isLoading, onLogin]);
  var onLogout = (0, _react.useCallback)( /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3() {
    return _regenerator["default"].wrap(function _callee3$(_context3) {
      while (1) switch (_context3.prev = _context3.next) {
        case 0:
          setIsLoading(true);
          _context3.prev = 1;
          _context3.next = 4;
          return provider.logout();
        case 4:
          _context3.next = 9;
          break;
        case 6:
          _context3.prev = 6;
          _context3.t0 = _context3["catch"](1);
          setError(_context3.t0);
        case 9:
          setIsLoading(false);
          setUser(null);
          setProvider(null);
        case 12:
        case "end":
          return _context3.stop();
      }
    }, _callee3, null, [[1, 6]]);
  })), [provider, setProvider]);
  var displayName = provider.displayName,
    name = provider.name;
  return /*#__PURE__*/_react["default"].createElement(StyledBox, null, provider.isNew ? /*#__PURE__*/_react["default"].createElement(NewTag, null, "New") : null, /*#__PURE__*/_react["default"].createElement("div", null), /*#__PURE__*/_react["default"].createElement(StyledTileWrapper, {
    onClick: onSelect,
    selected: isSelected
  }, /*#__PURE__*/_react["default"].createElement(StyledCloudName, null, displayName || name), provider.icon ? /*#__PURE__*/_react["default"].createElement(provider.icon, {
    height: "64px"
  }) : null, isLoading ? /*#__PURE__*/_react["default"].createElement("div", null, "Loading ...") : /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, user ? /*#__PURE__*/_react["default"].createElement(StyledUserName, null, actionName || user.name) : null), isSelected ? /*#__PURE__*/_react["default"].createElement(_styledComponents2.CheckMark, null) : null), user || error ? /*#__PURE__*/_react["default"].createElement(LogoutButton, {
    onClick: onLogout
  }) : /*#__PURE__*/_react["default"].createElement(LoginButton, {
    onClick: onLogin
  }), error ? /*#__PURE__*/_react["default"].createElement(StyledWarning, null, error.message) : null);
};
var _default = exports["default"] = CloudTile;
