// kepler.gl/dist/dist/side-panel/common/source-data-selector-content.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _react = _interopRequireWildcard(require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/node_modules/react"));
var _itemSelector = _interopRequireDefault(require("../../common/item-selector/item-selector"));
var _datasetTag = _interopRequireDefault(require("./dataset-tag"));
function _getRequireWildcardCache(e) {
  if ("function" != typeof WeakMap) return null;
  var r = new WeakMap(),
    t = new WeakMap();
  return (_getRequireWildcardCache = function _getRequireWildcardCache(e) {
    return e ? t : r;
  })(e);
}
function _interopRequireWildcard(e, r) {
  if (!r && e && e.__esModule) return e;
  if (null === e || "object" != _typeof(e) && "function" != typeof e) return {
    "default": e
  };
  var t = _getRequireWildcardCache(r);
  if (t && t.has(e)) return t.get(e);
  var n = {
      __proto__: null
    },
    a = Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) {
    var i = a ? Object.getOwnPropertyDescriptor(e, u) : null;
    i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u];
  }
  return n["default"] = e, t && t.set(e, n), n;
}
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

SourceDataSelectorContentFactory.deps = [_datasetTag["default"]];
function SourceDataSelectorContentFactory(DatasetTag) {
  var DatasetItem = function DatasetItem(_ref) {
    var value = _ref.value;
    return /*#__PURE__*/_react["default"].createElement(DatasetTag, {
      dataset: value
    });
  };
  var SourceDataSelectorContent = function SourceDataSelectorContent(_ref2) {
    var className = _ref2.className,
      datasets = _ref2.datasets,
      dataId = _ref2.dataId,
      inputTheme = _ref2.inputTheme,
      onSelect = _ref2.onSelect,
      defaultValue = _ref2.defaultValue,
      disabled = _ref2.disabled;
    var dsOptions = (0, _react.useMemo)(function () {
      return Object.values(datasets).map(function (ds) {
        return {
          label: ds.label,
          value: ds.id,
          color: ds.color
        };
      });
    }, [datasets]);
    var selectedItems = (0, _react.useMemo)(function () {
      return dataId ? (Array.isArray(dataId) && dataId || [dataId]).map(function (id) {
        return datasets[id];
      }) : [];
    }, [dataId, datasets]);
    return /*#__PURE__*/_react["default"].createElement(_itemSelector["default"], {
      className: className,
      inputTheme: inputTheme,
      selectedItems: selectedItems,
      options: dsOptions,
      getOptionValue: 'value',
      filterOption: 'label',
      multiSelect: false,
      onChange: onSelect,
      placeholder: defaultValue,
      disabled: disabled,
      displayOption: 'label',
      DropDownLineItemRenderComponent: DatasetItem
    });
  };
  return SourceDataSelectorContent;
}
var _default = exports["default"] = SourceDataSelectorContentFactory;
