// kepler.gl/dist/dist/side-panel/common/source-data-selector.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = SourceDataSelectorFactory;
var _react = _interopRequireDefault(require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/node_modules/react"));
var _localization = require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/src/localization/src");
var _styledComponents = require("../../common/styled-components");
var _sourceDataSelectorContent = _interopRequireDefault(require("./source-data-selector-content"));
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

SourceDataSelectorFactory.deps = [_sourceDataSelectorContent["default"]];
function SourceDataSelectorFactory(DataSourceSelectorContent) {
  var SourceDataSelector = /*#__PURE__*/_react["default"].memo(function (_ref) {
    var dataId = _ref.dataId,
      datasets = _ref.datasets,
      disabled = _ref.disabled,
      onSelect = _ref.onSelect,
      _ref$defaultValue = _ref.defaultValue,
      defaultValue = _ref$defaultValue === void 0 ? 'Select A Dataset' : _ref$defaultValue,
      inputTheme = _ref.inputTheme;
    return /*#__PURE__*/_react["default"].createElement(_styledComponents.SidePanelSection, {
      className: "data-source-selector"
    }, /*#__PURE__*/_react["default"].createElement(_styledComponents.PanelLabel, null, /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
      id: 'misc.dataSource'
    })), /*#__PURE__*/_react["default"].createElement(DataSourceSelectorContent, {
      inputTheme: inputTheme,
      datasets: datasets,
      dataId: dataId,
      onSelect: onSelect,
      defaultValue: defaultValue,
      disabled: disabled
    }));
  });
  SourceDataSelector.displayName = 'SourceDataSelector';
  return SourceDataSelector;
}
