// kepler.gl/dist/dist/common/animation-control/play-control.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/node_modules/react"));
var _classnames = _interopRequireDefault(require("classnames"));
var _localization = require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/src/localization/src");
var _iconButton = _interopRequireDefault(require("../icon-button"));
var _tippyTooltip = _interopRequireDefault(require("../tippy-tooltip"));
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

function PlayControlFactory() {
  var PlayControl = function PlayControl(_ref) {
    var showAnimationWindowControl = _ref.showAnimationWindowControl,
      isAnimating = _ref.isAnimating,
      pauseAnimation = _ref.pauseAnimation,
      startAnimation = _ref.startAnimation,
      isSpeedControlVisible = _ref.isSpeedControlVisible,
      btnStyle = _ref.btnStyle,
      playbackIcons = _ref.playbackIcons,
      buttonHeight = _ref.buttonHeight;
    return showAnimationWindowControl ? null : /*#__PURE__*/_react["default"].createElement(_tippyTooltip["default"], {
      placement: "top",
      delay: [500, 0],
      render: function render() {
        return /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
          id: isAnimating ? 'tooltip.pause' : 'tooltip.play'
        });
      }
    }, /*#__PURE__*/_react["default"].createElement(_iconButton["default"], (0, _extends2["default"])({
      className: (0, _classnames["default"])('playback-control-button', {
        active: isAnimating
      }),
      onClick: isAnimating ? pauseAnimation : startAnimation,
      hide: isSpeedControlVisible
    }, btnStyle), isAnimating ? /*#__PURE__*/_react["default"].createElement(playbackIcons.pause, {
      height: buttonHeight
    }) : /*#__PURE__*/_react["default"].createElement(playbackIcons.play, {
      height: buttonHeight
    })));
  };
  return PlayControl;
}
var _default = exports["default"] = PlayControlFactory;
