// kepler.gl/src/ai-assistant/dist/components/ai-assistant-manager.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AiAssistantPanel = AiAssistantPanel;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));
var _react = _interopRequireWildcard(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _reactRedux = require("react-redux");
var _reactIntl = require("react-intl");
var _utils = require("@kepler.gl/utils");
var _localization = require("@kepler.gl/localization");
var _components = require("@kepler.gl/components");
var _actions = require("../actions");
var _aiAssistantConfig = require("./ai-assistant-config");
var _aiAssistantComponent = require("./ai-assistant-component");
var _localization2 = require("../localization");
var _templateObject, _templateObject2, _templateObject3, _templateObject4; // SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2["default"])(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StyledAiAssistantPanelContainer = _styledComponents["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  flex-direction: column;\n  pointer-events: none !important; /* prevent padding from blocking input */\n  flex-grow: 1;\n  justify-content: space-between;\n  overflow: hidden;\n  height: 100%;\n  width: 100%;\n  & > * {\n    /* all children should allow input */\n    pointer-events: all;\n  }\n"])));
var StyledAiAssistantPanel = _styledComponents["default"].div(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  top: 0;\n  background-color: ", ";\n  display: flex;\n  flex-direction: column;\n  flex-grow: 1;\n  overflow: hidden;\n"])), function (props) {
  return props.theme.sidePanelBg;
});
var StyledAiAssistantPanelHeader = _styledComponents["default"].div(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2["default"])(["\n  padding: 16px 16px 4px 16px;\n  border-bottom: 1px solid ", ";\n  color: ", ";\n"])), function (props) {
  return props.theme.borderColor;
}, function (props) {
  return props.theme.subtextColorActive;
});
var StyledAiAssistantPanelContent = _styledComponents["default"].div(_templateObject4 || (_templateObject4 = (0, _taggedTemplateLiteral2["default"])(["\n  ", ";\n  color: ", ";\n  padding: 10px 0px 10px 0px;\n  overflow-y: auto;\n  display: flex;\n  flex-direction: column;\n  height: 100%;\n"])), function (props) {
  return props.theme.sidePanelScrollBar;
}, function (props) {
  return props.theme.subtextColorActive;
});
var SidePanelTitle = (0, _components.SidePanelTitleFactory)();
function AiAssistantPanel() {
  var dispatch = (0, _reactRedux.useDispatch)();
  var aiAssistant = (0, _reactRedux.useSelector)(function (state) {
    return state.demo.aiAssistant;
  });
  var locale = (0, _reactRedux.useSelector)(function (state) {
    return state.demo.keplerGl.map.uiState.locale;
  });
  var onConfigButtonClick = (0, _react.useCallback)(function () {
    if (aiAssistant) {
      // set aiAssistant.config.isReady to false so we can render the config component
      dispatch((0, _actions.updateAiAssistantConfig)(_objectSpread(_objectSpread({}, aiAssistant.config), {}, {
        isReady: false
      })));
    }
  }, [aiAssistant, dispatch]);

  // combine keplerGlMessages and messages
  var combinedMessages = (0, _react.useMemo)(function () {
    return Object.keys(_localization2.messages).reduce(function (acc, language) {
      return _objectSpread(_objectSpread({}, acc), {}, (0, _defineProperty2["default"])({}, language, _objectSpread(_objectSpread({}, _localization2.messages[language] || {}), _localization.messages[language] || {})));
    }, {});
  }, []);
  return /*#__PURE__*/_react["default"].createElement(_reactIntl.IntlProvider, {
    locale: locale,
    messages: (0, _utils.flattenMessages)(combinedMessages[locale])
  }, /*#__PURE__*/_react["default"].createElement(StyledAiAssistantPanelContainer, {
    className: "ai-assistant-manager"
  }, /*#__PURE__*/_react["default"].createElement(StyledAiAssistantPanel, null, /*#__PURE__*/_react["default"].createElement(StyledAiAssistantPanelHeader, null, /*#__PURE__*/_react["default"].createElement(SidePanelTitle, {
    className: "ai-assistant-manager-title",
    title: "AI Assistant"
  }, /*#__PURE__*/_react["default"].createElement(_components.Icons.Settings, {
    onClick: onConfigButtonClick
  }))), /*#__PURE__*/_react["default"].createElement(StyledAiAssistantPanelContent, null, !(aiAssistant !== null && aiAssistant !== void 0 && aiAssistant.config.isReady) ? /*#__PURE__*/_react["default"].createElement(_aiAssistantConfig.AiAssistantConfig, null) : /*#__PURE__*/_react["default"].createElement(_aiAssistantComponent.AiAssistantComponent, null)))));
}
