// kepler.gl/dist/dist/bottom-widget.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = BottomWidgetFactory;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));
var _react = _interopRequireWildcard(require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/node_modules/react"));
var _styledComponents = _interopRequireWildcard(require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/node_modules/styled-components"));
var _constants = require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/src/constants/src");
var _utils = require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/src/utils/src");
var _styles = require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/src/styles/src");
var _timeWidget = _interopRequireDefault(require("./filters/time-widget"));
var _filterAnimationController = _interopRequireDefault(require("./filter-animation-controller"));
var _layerAnimationController = _interopRequireDefault(require("./layer-animation-controller"));
var _animationControl = _interopRequireDefault(require("./common/animation-control/animation-control"));
var _templateObject, _templateObject2, _templateObject3; // SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
function _getRequireWildcardCache(e) {
  if ("function" != typeof WeakMap) return null;
  var r = new WeakMap(),
    t = new WeakMap();
  return (_getRequireWildcardCache = function _getRequireWildcardCache(e) {
    return e ? t : r;
  })(e);
}
function _interopRequireWildcard(e, r) {
  if (!r && e && e.__esModule) return e;
  if (null === e || "object" != _typeof(e) && "function" != typeof e) return {
    "default": e
  };
  var t = _getRequireWildcardCache(r);
  if (t && t.has(e)) return t.get(e);
  var n = {
      __proto__: null
    },
    a = Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) {
    var i = a ? Object.getOwnPropertyDescriptor(e, u) : null;
    i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u];
  }
  return n["default"] = e, t && t.set(e, n), n;
}
var maxWidth = 1080;
var BottomWidgetContainer = _styledComponents["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  flex-direction: column;\n  padding-top: ", "px;\n  padding-right: ", "px;\n  padding-bottom: ", "px;\n  padding-left: ", "px;\n  pointer-events: none !important; /* prevent padding from blocking input */\n  & > * {\n    /* all children should allow input */\n    pointer-events: all;\n  }\n  width: ", "px;\n  z-index: 1;\n  ", "\n"])), function (props) {
  return props.hasPadding ? props.theme.bottomWidgetPaddingTop : 0;
}, function (props) {
  return props.hasPadding ? props.theme.bottomWidgetPaddingRight : 0;
}, function (props) {
  return props.hasPadding ? props.theme.bottomWidgetPaddingBottom : 0;
}, function (props) {
  return props.hasPadding ? props.theme.bottomWidgetPaddingLeft : 0;
}, function (props) {
  return props.width;
}, _styles.media.portable(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["padding: 0;"]))));
BottomWidgetFactory.deps = [_timeWidget["default"], _animationControl["default"], _filterAnimationController["default"], _layerAnimationController["default"]];

/* eslint-disable complexity */
function BottomWidgetFactory(TimeWidget, AnimationControl, FilterAnimationController, LayerAnimationController) {
  var LayerAnimationControl = (0, _styledComponents["default"])(AnimationControl)(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2["default"])(["\n    background-color: ", ";\n  "])), function (props) {
    return props.theme.sidePanelBg;
  });
  var BottomWidget = function BottomWidget(props) {
    var _uiState$mapControls, _uiState$mapControls2, _theme$mapControl, _theme$mapControl2;
    var datasets = props.datasets,
      filters = props.filters,
      animationConfig = props.animationConfig,
      visStateActions = props.visStateActions,
      containerW = props.containerW,
      uiState = props.uiState,
      sidePanelWidth = props.sidePanelWidth,
      layers = props.layers,
      rootRef = props.rootRef,
      theme = props.theme;
    var activeSidePanel = uiState.activeSidePanel,
      readOnly = uiState.readOnly;
    var isOpen = Boolean(activeSidePanel);
    var enlargedFilterIdx = (0, _react.useMemo)(function () {
      return filters.findIndex(function (f) {
        return !(0, _utils.isSideFilter)(f);
      });
    }, [filters]);
    var animatedFilterIdx = (0, _react.useMemo)(function () {
      return filters.findIndex(function (f) {
        return f.isAnimating;
      });
    }, [filters]);
    var animatedFilter = animatedFilterIdx > -1 ? filters[animatedFilterIdx] : null;
    // we need to hide layer timeline when filter is synced and not enlarged
    var isTimelineLinkedWithFilter = (0, _react.useMemo)(function () {
      return filters.some(function (f) {
        return f.syncedWithLayerTimeline;
      });
    }, [filters]);
    var isMobile = (0, _react.useMemo)(function () {
      return (0, _utils.hasPortableWidth)(_styles.breakPointValues);
    }, []);
    var isLegendPinned = ((_uiState$mapControls = uiState.mapControls) === null || _uiState$mapControls === void 0 || (_uiState$mapControls = _uiState$mapControls.mapLegend) === null || _uiState$mapControls === void 0 ? void 0 : _uiState$mapControls.show) && ((_uiState$mapControls2 = uiState.mapControls) === null || _uiState$mapControls2 === void 0 || (_uiState$mapControls2 = _uiState$mapControls2.mapLegend) === null || _uiState$mapControls2 === void 0 ? void 0 : _uiState$mapControls2.active);
    var spaceForLegendWidth = isLegendPinned ? ((_theme$mapControl = theme.mapControl) === null || _theme$mapControl === void 0 ? void 0 : _theme$mapControl.width) + ((_theme$mapControl2 = theme.mapControl) === null || _theme$mapControl2 === void 0 || (_theme$mapControl2 = _theme$mapControl2.mapLegend) === null || _theme$mapControl2 === void 0 || (_theme$mapControl2 = _theme$mapControl2.pinned) === null || _theme$mapControl2 === void 0 ? void 0 : _theme$mapControl2.right) * 2 - theme.bottomWidgetPaddingRight : 0;
    var enlargedFilterWidth = (0, _react.useMemo)(function () {
      return (!isMobile && isOpen ? containerW - sidePanelWidth : containerW) - spaceForLegendWidth;
    }, [isMobile, isOpen, containerW, sidePanelWidth, spaceForLegendWidth]);

    // show playback control if layers contain trip layer & at least one trip layer is visible
    var animatableLayer = (0, _react.useMemo)(function () {
      return layers.filter(function (l) {
        return l.config.animation && l.config.animation.enabled && l.config.isVisible;
      });
    }, [layers]);
    var readyToAnimation = (0, _react.useMemo)(function () {
      return Array.isArray(animationConfig.domain) && Number.isFinite(animationConfig.currentTime);
    }, [animationConfig.domain, animationConfig.currentTime]);

    // if animation control is showing, hide time display in time slider
    var showFloatingTimeDisplay = !animatableLayer.length;
    var showAnimationControl = animatableLayer.length && readyToAnimation && !animationConfig.hideControl;
    var showTimeWidget = enlargedFilterIdx > -1 && Object.keys(datasets).length > 0;

    // if filter is not animating, pass in enlarged filter here because
    // animation controller needs to call reset on it
    var filter = (0, _react.useMemo)(function () {
      return animatedFilter || filters[enlargedFilterIdx];
    }, [animatedFilter, filters, enlargedFilterIdx]);

    // we merge filter and timeline if filter is synced
    var _useMemo = (0, _react.useMemo)(function () {
        return filter !== null && filter !== void 0 && filter.syncedWithLayerTimeline ? (0, _utils.mergeFilterWithTimeline)(filter, animationConfig) : {
          filter: filter,
          animationConfig: animationConfig
        };
      }, [filter, animationConfig]),
      enhancedFilter = _useMemo.filter,
      enhancedAnimationConfig = _useMemo.animationConfig;
    var onClose = (0, _react.useCallback)(function () {
      return visStateActions.setFilterView(enlargedFilterIdx, _constants.FILTER_VIEW_TYPES.side);
    }, [visStateActions, enlargedFilterIdx]);
    var onToggleMinify = (0, _react.useCallback)(function () {
      return visStateActions.setFilterView(enlargedFilterIdx, filter.view === _constants.FILTER_VIEW_TYPES.enlarged ? _constants.FILTER_VIEW_TYPES.minified : _constants.FILTER_VIEW_TYPES.enlarged);
    }, [enlargedFilterIdx, visStateActions, filter]);
    return /*#__PURE__*/_react["default"].createElement(BottomWidgetContainer, {
      width: Math.min(maxWidth, enlargedFilterWidth),
      style: {
        marginRight: spaceForLegendWidth
      },
      className: "bottom-widget--container",
      hasPadding: showAnimationControl || showTimeWidget,
      ref: rootRef
    }, !isTimelineLinkedWithFilter ? /*#__PURE__*/_react["default"].createElement(LayerAnimationController, {
      animationConfig: enhancedAnimationConfig,
      setLayerAnimationTime: visStateActions.setLayerAnimationTime
    }, function (isAnimating, start, pause, resetAnimation, timeline, setTimelineValue) {
      return showAnimationControl ? /*#__PURE__*/_react["default"].createElement(LayerAnimationControl, {
        updateAnimationSpeed: visStateActions.updateLayerAnimationSpeed,
        toggleAnimation: visStateActions.toggleLayerAnimation,
        isAnimatable: !animatedFilter,
        isAnimating: isAnimating,
        resetAnimation: resetAnimation,
        setTimelineValue: setTimelineValue,
        timeline: timeline
      }) : null;
    }) : null, enhancedFilter ? /*#__PURE__*/_react["default"].createElement(FilterAnimationController, {
      filter: enhancedFilter,
      filterIdx: animatedFilterIdx > -1 ? animatedFilterIdx : enlargedFilterIdx,
      setFilterAnimationTime: visStateActions.setFilterAnimationTime
    }, function (isAnimating, start, pause, resetAnimation, timeline, setTimelineValue) {
      return showTimeWidget && timeline ? /*#__PURE__*/_react["default"].createElement(TimeWidget
      // TimeWidget uses React.memo, here we pass width
      // even though it doesnt use it, to force rerender
      , {
        filter: enhancedFilter,
        index: enlargedFilterIdx,
        datasets: datasets,
        layers: layers,
        readOnly: readOnly,
        showTimeDisplay: showFloatingTimeDisplay,
        setFilterPlot: visStateActions.setFilterPlot,
        setFilterAnimationTime: setTimelineValue,
        setFilterAnimationWindow: visStateActions.setFilterAnimationWindow,
        setFilterSyncTimelineMode: visStateActions.setTimeFilterSyncTimelineMode,
        toggleAnimation: visStateActions.toggleFilterAnimation,
        updateAnimationSpeed: visStateActions.updateFilterAnimationSpeed,
        resetAnimation: resetAnimation,
        isAnimatable: !animationConfig || !animationConfig.isAnimating,
        animationConfig: animationConfig,
        onClose: onClose,
        timeline: timeline,
        onToggleMinify: onToggleMinify
      }) : null;
    }) : null);
  };
  return (0, _styledComponents.withTheme)( /*#__PURE__*/(0, _react.forwardRef)(function (props, ref) {
    return /*#__PURE__*/_react["default"].createElement(BottomWidget, (0, _extends2["default"])({}, props, {
      rootRef: ref
    }));
  }));
}
/* eslint-enable complexity */
