// kepler.gl/src/duckdb/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  DuckDBWasmAdapter: true,
  SqlPanel: true,
  restoreGeoarrowMetadata: true,
  KeplerGlDuckDbTable: true,
  consolidateFieldTypes: true,
  extendFieldsWithDuckDBColumnType: true,
  fieldTypeToColumnType: true
};
Object.defineProperty(exports, "DuckDBWasmAdapter", {
  enumerable: true,
  get: function get() {
    return _duckdbWasmAdapter.DuckDBWasmAdapter;
  }
});
Object.defineProperty(exports, "KeplerGlDuckDbTable", {
  enumerable: true,
  get: function get() {
    return _duckdbTable.KeplerGlDuckDbTable;
  }
});
Object.defineProperty(exports, "SqlPanel", {
  enumerable: true,
  get: function get() {
    return _components.SqlPanel;
  }
});
Object.defineProperty(exports, "consolidateFieldTypes", {
  enumerable: true,
  get: function get() {
    return _dataProcessor.consolidateFieldTypes;
  }
});
exports["default"] = void 0;
Object.defineProperty(exports, "extendFieldsWithDuckDBColumnType", {
  enumerable: true,
  get: function get() {
    return _dataProcessor.extendFieldsWithDuckDBColumnType;
  }
});
Object.defineProperty(exports, "fieldTypeToColumnType", {
  enumerable: true,
  get: function get() {
    return _dataProcessor.fieldTypeToColumnType;
  }
});
Object.defineProperty(exports, "restoreGeoarrowMetadata", {
  enumerable: true,
  get: function get() {
    return _duckdbTable.restoreGeoarrowMetadata;
  }
});
var _duckdbWasmAdapter = require("./adapters/duckdb-wasm-adapter");
var _components = require("./components");
var _plugin = require("./plugin");
var _duckdbTable = require("./table/duckdb-table");
var _duckdbTableUtils = require("./table/duckdb-table-utils");
Object.keys(_duckdbTableUtils).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _duckdbTableUtils[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _duckdbTableUtils[key];
    }
  });
});
var _dataProcessor = require("./processors/data-processor");
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
var _default = exports["default"] = _plugin.keplerGlDuckDBPlugin;
