// kepler.gl/src/ai-assistant/dist/tools/query-tool.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getQueryTool = getQueryTool;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _react = _interopRequireDefault(require("react"));
var _duckdb = require("@openassistant/duckdb");
var _tables = require("@openassistant/tables");
var _utils = require("@openassistant/utils");
var _utils2 = require("./utils");
var _tableTool = require("./kepler-tools/table-tool");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2["default"])(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; } // SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
function getQueryTool(datasets, layers) {
  var toolCache = _utils.ToolCache.getInstance();

  // context for query tools
  var getValues = /*#__PURE__*/function () {
    var _ref = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(datasetName, variableName) {
      var values;
      return _regenerator["default"].wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            values = (0, _utils2.getValuesFromDataset)(datasets, layers, datasetName, variableName);
            return _context.abrupt("return", values);
          case 2:
          case "end":
            return _context.stop();
        }
      }, _callee);
    }));
    return function getValues(_x, _x2) {
      return _ref.apply(this, arguments);
    };
  }();

  // customize some query tools from localQuery tool
  function QueryToolComponent(props) {
    return /*#__PURE__*/_react["default"].createElement(_tables.QueryDuckDBComponent, (0, _extends2["default"])({}, props, {
      getValues: getValues,
      getDuckDB: _duckdb.getDuckDB
    }));
  }

  // this tool will execute a generic select SQL query against user's dataset
  var genericQuery = _objectSpread(_objectSpread({}, _duckdb.localQuery), {}, {
    description: "execute a generic select SQL query in duckdb to answer user's question. Please note:\n1. This tool is NOT for filtering the user dataset.\n2. This tool does NOT support geometry column and geometric operations.\n3. The variableNames should not be empty. If it is not provided, then pick a variable name from the dataset.\n4. There is no need to add a sub-query to add an auto-increment column 'row_index' to the original dataset.\n",
    context: _objectSpread(_objectSpread({}, _duckdb.localQuery.context), {}, {
      getValues: getValues
    }),
    component: QueryToolComponent
  });

  // customize a filterDataset tool from localQuery tool
  // this tool will use the selected rows (row indexes) to filter the user dataset and save the result as a new dataset.
  var filterDataset = _objectSpread(_objectSpread({}, _duckdb.localQuery), {}, {
    description: "filter the user dataset by using a select SQL query in duckdb and save the result as a new dataset.\nPlease note:\n1. Do not use * to select all columns, instead use all the column names in dataset.\n",
    context: _objectSpread(_objectSpread({}, _duckdb.localQuery.context), {}, {
      getValues: getValues
    }),
    component: _tableTool.TableToolComponent
  });

  // customize a table tool from localQuery tool to create a new table/dataset in kepler.gl using SQL query
  var tableTool = _objectSpread(_objectSpread({}, _duckdb.localQuery), {}, {
    description: "Create a new table/dataset in kepler.gl using SQL query which will\n1. add a new column to the original dataset\n2. delete a column from the original dataset\n3. rename a column in the original dataset\n4. change the column type in the original dataset.\nPlease note:\n1. Do not use * to select all columns, instead use all the column names in dataset.\n2. List all column names the new table or dataset will have.\n",
    context: _objectSpread(_objectSpread({}, _duckdb.localQuery.context), {}, {
      getValues: getValues
    }),
    component: _tableTool.TableToolComponent
  });

  // customize a mergeTables tool from localQuery tool to merge two datasets into a new dataset
  var mergeTablesTool = _objectSpread(_objectSpread({}, _duckdb.mergeTables), {}, {
    context: {
      getValues: getValues
    },
    onToolCompleted: function onToolCompleted(toolName, additionalData) {
      toolCache.addDataset(toolName, additionalData);
    }
    // component: MergeTablesToolComponent
  });
  return {
    filterDataset: filterDataset,
    genericQuery: genericQuery,
    tableTool: tableTool,
    mergeTablesTool: mergeTablesTool
  };
}
