// kepler.gl/dist/dist/side-panel/interaction-manager.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _react = _interopRequireDefault(require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/node_modules/react"));
var _reactIntl = require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/node_modules/react-intl");
var _interactionPanel = _interopRequireDefault(require("./interaction-panel/interaction-panel"));
var _panelTitle = _interopRequireDefault(require("./panel-title"));
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

InteractionManagerFactory.deps = [_interactionPanel["default"], _panelTitle["default"]];
function InteractionManagerFactory(InteractionPanel, PanelTitle) {
  var InteractionManager = function InteractionManager(_ref) {
    var interactionConfig = _ref.interactionConfig,
      datasets = _ref.datasets,
      visStateActions = _ref.visStateActions,
      panelMetadata = _ref.panelMetadata;
    var onConfigChange = visStateActions.interactionConfigChange,
      setColumnDisplayFormat = visStateActions.setColumnDisplayFormat;
    var intl = (0, _reactIntl.useIntl)();
    return /*#__PURE__*/_react["default"].createElement("div", {
      className: "interaction-manager"
    }, /*#__PURE__*/_react["default"].createElement(PanelTitle, {
      className: "interaction-manager-title",
      title: intl.formatMessage({
        id: panelMetadata.label
      })
    }), Object.keys(interactionConfig).map(function (key) {
      return /*#__PURE__*/_react["default"].createElement(InteractionPanel, {
        datasets: datasets,
        config: interactionConfig[key],
        key: key,
        onConfigChange: onConfigChange,
        setColumnDisplayFormat: setColumnDisplayFormat
      });
    }));
  };
  return InteractionManager;
}
var _default = exports["default"] = InteractionManagerFactory;
