// @hubble.gl/react/dist/components/resolution-guide.js
// hubble.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
import React from 'react';
export default function ResolutionGuide() {
    return (React.createElement("div", { style: { position: 'relative' } },
        React.createElement(Outline, { width: 2560, height: 1440, name: '2160p, 2K' }),
        React.createElement(Outline, { width: 2560, height: 1440, name: '1440p' }),
        React.createElement(Outline, { width: 1920, height: 1080, name: '1080p, Full HD' }),
        React.createElement(Outline, { width: 1280, height: 720, name: '720p, HD' }),
        React.createElement(Outline, { width: 854, height: 480, name: '480p, WSD' }),
        React.createElement(Outline, { width: 640, height: 480, name: '480p, SD' }),
        React.createElement(Outline, { width: 640, height: 360, name: '360p' }),
        React.createElement(Outline, { width: 426, height: 240, name: '240p' })));
}
function Outline({ width, height, name }) {
    return (React.createElement("div", { style: { position: 'absolute', width, height, outline: '2px dashed black' } },
        React.createElement("div", { style: {
                position: 'absolute',
                bottom: -20,
                right: 0,
                fontSize: '14px',
                font: "normal 14px/20px 'Uber Move',Helvetica,Arial,sans-serif"
            } },
            name,
            " (",
            width,
            " x ",
            height,
            ")")));
}
