// kepler.gl/src/duckdb/dist/components/schema-panel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SchemaPanelDropMessage = exports.SchemaPanel = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));
var _react = _interopRequireWildcard(require("react"));
var _reactRedux = require("react-redux");
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _components = require("@kepler.gl/components");
var _processors = require("@kepler.gl/processors");
var _utils = require("@kepler.gl/utils");
var _tree = require("./tree");
var _duckdbTableUtils = require("../table/duckdb-table-utils");
var _templateObject, _templateObject2, _templateObject3, _templateObject4; // SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != _typeof(e) && "function" != typeof e) return { "default": e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n["default"] = e, t && t.set(e, n), n; }
// TODO note that demo state is available in demo-app, but not when add modules to dependencies in a custom map

var StyledSchemaPanel = _styledComponents["default"].div(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  color: ", ";\n  font-size: 12px;\n  padding: 12px;\n  font-family: ", ";\n  height: 100%;\n"])), function (props) {
  return props.theme.textColor;
}, function (props) {
  return props.theme.fontFamily;
});
var StyledLoadingSpinnerWrapper = _styledComponents["default"].div(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  justify-content: center;\n  align-items: center;\n  height: 100%;\n"])));
function getColumnSchema(_x, _x2) {
  return _getColumnSchema.apply(this, arguments);
}
function _getColumnSchema() {
  _getColumnSchema = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(connection, tableName) {
    var columnResult, columnDescribe, keplerFields;
    return _regenerator["default"].wrap(function _callee2$(_context2) {
      while (1) switch (_context2.prev = _context2.next) {
        case 0:
          _context2.next = 2;
          return connection.query("Select * from '".concat(tableName, "' LIMIT 1;"));
        case 2:
          columnResult = _context2.sent;
          _context2.next = 5;
          return (0, _duckdbTableUtils.getDuckDBColumnTypes)(connection, tableName);
        case 5:
          columnDescribe = _context2.sent;
          keplerFields = (0, _processors.arrowSchemaToFields)(columnResult, (0, _duckdbTableUtils.getDuckDBColumnTypesMap)(columnDescribe));
          return _context2.abrupt("return", {
            key: tableName,
            object: {
              type: 'dataset',
              tableName: tableName
            },
            children: columnResult.schema.fields.map(function (field, fieldIndex) {
              return {
                key: field.name,
                object: {
                  type: 'column',
                  name: field.name,
                  arrowType: field.type,
                  fieldType: keplerFields[fieldIndex].type
                }
              };
            })
          });
        case 8:
        case "end":
          return _context2.stop();
      }
    }, _callee2);
  }));
  return _getColumnSchema.apply(this, arguments);
}
function getSchemaSuggestion(result) {
  return result.reduce(function (accu, data) {
    var columns = data.children.map(function (child) {
      return {
        column_name: child.key,
        table_name: data.key
      };
    });
    return accu.concat(columns);
  }, []);
}
var StyledSchemaPanelDropMessage = _styledComponents["default"].div(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2["default"])(["\n  display: flex;\n  justify-content: center;\n  align-items: center;\n  height: 100%;\n  flex-direction: column;\n  text-align: center;\n\n  div {\n    margin: 5px;\n  }\n  .header {\n    font-size: 15px;\n  }\n  .bold {\n    font-weight: 700;\n  }\n"])));
var StyledAddIcon = (0, _styledComponents["default"])(_components.Icons.Add)(_templateObject4 || (_templateObject4 = (0, _taggedTemplateLiteral2["default"])(["\n  display: inline;\n  margin-top: -3px;\n"])));
var SchemaPanelDropMessage = exports.SchemaPanelDropMessage = function SchemaPanelDropMessage() {
  return /*#__PURE__*/_react["default"].createElement(StyledSchemaPanelDropMessage, null, /*#__PURE__*/_react["default"].createElement("div", {
    className: "header"
  }, /*#__PURE__*/_react["default"].createElement(StyledAddIcon, null), " Add files to DuckDB"), /*#__PURE__*/_react["default"].createElement("div", {
    className: "bold"
  }, "Supported formats: "), /*#__PURE__*/_react["default"].createElement("div", null, ".csv, .json, .geojson, .parquet, .arrow"), /*#__PURE__*/_react["default"].createElement("div", null, "Files you add will stay local to your browser."));
};
var SchemaPanel = exports.SchemaPanel = function SchemaPanel(_ref) {
  var setTableSchema = _ref.setTableSchema,
    droppedFile = _ref.droppedFile,
    schemaUpdateTrigger = _ref.schemaUpdateTrigger;
  var _useState = (0, _react.useState)([]),
    _useState2 = (0, _slicedToArray2["default"])(_useState, 2),
    columnSchemas = _useState2[0],
    setColumnSchemas = _useState2[1];
  var datasets = (0, _reactRedux.useSelector)(function (state) {
    var _state$demo;
    return state === null || state === void 0 || (_state$demo = state.demo) === null || _state$demo === void 0 || (_state$demo = _state$demo.keplerGl) === null || _state$demo === void 0 || (_state$demo = _state$demo.map) === null || _state$demo === void 0 ? void 0 : _state$demo.visState.datasets;
  });
  var getTableSchema = (0, _react.useCallback)( /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
    var _tableResult$getChild, _ref3;
    var db, c, tableResult, tableNames, result, tableSchema;
    return _regenerator["default"].wrap(function _callee$(_context) {
      while (1) switch (_context.prev = _context.next) {
        case 0:
          db = (0, _utils.getApplicationConfig)().database;
          if (db) {
            _context.next = 4;
            break;
          }
          console.error('The database is not configured properly.');
          return _context.abrupt("return");
        case 4:
          _context.next = 6;
          return db.connect();
        case 6:
          c = _context.sent;
          _context.next = 9;
          return c.query('SHOW TABLES;');
        case 9:
          tableResult = _context.sent;
          tableNames = (_tableResult$getChild = tableResult.getChildAt(0)) === null || _tableResult$getChild === void 0 ? void 0 : _tableResult$getChild.toJSON();
          _context.next = 13;
          return Promise.all((_ref3 = tableNames || []) === null || _ref3 === void 0 ? void 0 : _ref3.map(function (name) {
            return getColumnSchema(c, name);
          }));
        case 13:
          result = _context.sent;
          tableSchema = getSchemaSuggestion(result);
          setColumnSchemas(result);
          setTableSchema(tableSchema);
          _context.next = 19;
          return c.close();
        case 19:
        case "end":
          return _context.stop();
      }
    }, _callee);
  })), [setColumnSchemas, setTableSchema, schemaUpdateTrigger]);
  (0, _react.useEffect)(function () {
    getTableSchema();
  }, [datasets, droppedFile, getTableSchema]);
  return /*#__PURE__*/_react["default"].createElement(StyledSchemaPanel, null, columnSchemas !== null && columnSchemas !== void 0 && columnSchemas.length ? columnSchemas.map(function (data) {
    return /*#__PURE__*/_react["default"].createElement(_tree.Tree, {
      key: data.key,
      treeData: data,
      renderNode: function renderNode(node) {
        if (node.object.type === 'dataset') {
          return /*#__PURE__*/_react["default"].createElement(_tree.DatasetNode, {
            node: node
          });
        } else if (node.object.type === 'column') {
          return /*#__PURE__*/_react["default"].createElement(_tree.ColumnNode, {
            node: node
          });
        }
        return null;
      }
    });
  }) : droppedFile ? /*#__PURE__*/_react["default"].createElement(StyledLoadingSpinnerWrapper, null, /*#__PURE__*/_react["default"].createElement(_components.LoadingSpinner, null)) : /*#__PURE__*/_react["default"].createElement(SchemaPanelDropMessage, null));
};
