// kepler.gl/dist/dist/side-panel/layer-panel/aggr-scale-selector.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AggrScaleSelectorFactory = AggrScaleSelectorFactory;
exports["default"] = void 0;
var _react = _interopRequireDefault(require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/node_modules/react"));
var _dimensionScaleSelector = _interopRequireDefault(require("./dimension-scale-selector"));
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

AggrScaleSelectorFactory.deps = [_dimensionScaleSelector["default"]];
function AggrScaleSelectorFactory(DimensionScaleSelector) {
  var AggrScaleSelector = function AggrScaleSelector(_ref) {
    var channel = _ref.channel,
      dataset = _ref.dataset,
      layer = _ref.layer,
      onChange = _ref.onChange,
      setColorUI = _ref.setColorUI,
      label = _ref.label;
    var key = channel.key;
    var scaleOptions = layer.getScaleOptions(key);
    return Array.isArray(scaleOptions) && scaleOptions.length > 1 ? /*#__PURE__*/_react["default"].createElement(DimensionScaleSelector, {
      dataset: dataset,
      layer: layer,
      channel: channel,
      label: label || "".concat(key, " Scale"),
      onChange: onChange,
      setColorUI: setColorUI
    }) : null;
  };
  return AggrScaleSelector;
}
var _default = exports["default"] = AggrScaleSelectorFactory;
