// kepler.gl/dist/dist/side-panel/layer-panel/dimension-scale-selector.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _objectDestructuringEmpty2 = _interopRequireDefault(require("@babel/runtime/helpers/objectDestructuringEmpty"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _constants = require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/src/constants/src");
var _localization = require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/src/localization/src");
var _utils = require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/src/utils/src");
var _react = _interopRequireWildcard(require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/node_modules/react"));
var _itemSelector = _interopRequireDefault(require("../../common/item-selector/item-selector"));
var _styledComponents = require("../../common/styled-components");
var _colorScaleSelector = _interopRequireDefault(require("./color-scale-selector"));
function _getRequireWildcardCache(e) {
  if ("function" != typeof WeakMap) return null;
  var r = new WeakMap(),
    t = new WeakMap();
  return (_getRequireWildcardCache = function _getRequireWildcardCache(e) {
    return e ? t : r;
  })(e);
}
function _interopRequireWildcard(e, r) {
  if (!r && e && e.__esModule) return e;
  if (null === e || "object" != _typeof(e) && "function" != typeof e) return {
    "default": e
  };
  var t = _getRequireWildcardCache(r);
  if (t && t.has(e)) return t.get(e);
  var n = {
      __proto__: null
    },
    a = Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) {
    var i = a ? Object.getOwnPropertyDescriptor(e, u) : null;
    i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u];
  }
  return n["default"] = e, t && t.set(e, n), n;
}
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

var SizeScaleSelector = function SizeScaleSelector(_ref) {
  var dropdownSelectProps = (0, _extends2["default"])({}, ((0, _objectDestructuringEmpty2["default"])(_ref), _ref));
  return /*#__PURE__*/_react["default"].createElement(_itemSelector["default"], dropdownSelectProps);
};
DimensionScaleSelectorFactory.deps = [_colorScaleSelector["default"]];
function DimensionScaleSelectorFactory(ColorScaleSelector) {
  var DimensionScaleSelector = function DimensionScaleSelector(_ref2) {
    var _layer$config$colorUI;
    var layer = _ref2.layer,
      channel = _ref2.channel,
      dataset = _ref2.dataset,
      label = _ref2.label,
      onChange = _ref2.onChange,
      setColorUI = _ref2.setColorUI;
    var channelScaleType = channel.channelScaleType,
      domain = channel.domain,
      field = channel.field,
      key = channel.key,
      range = channel.range,
      scale = channel.scale;
    var scaleType = scale ? layer.config[scale] : null;
    var layerScaleOptions = layer.getScaleOptions(key);
    var scaleOptions = layerScaleOptions.map(function (op) {
      return {
        label: _constants.SCALE_TYPE_NAMES[op] || op,
        value: op
      };
    });
    var disabled = scaleOptions.length < 2;
    var isColorScale = channelScaleType === _constants.CHANNEL_SCALES.color || layer.config.aggregatedBins && channelScaleType === _constants.CHANNEL_SCALES.colorAggr;
    var onSelect = (0, _react.useCallback)(function (val, newRange) {
      return onChange((0, _defineProperty2["default"])({}, scale, val), key, newRange ? (0, _defineProperty2["default"])({}, range, newRange) : undefined);
    }, [onChange, range, scale, key]);
    var _setColorUI = (0, _react.useCallback)(function (newConfig) {
      return setColorUI(range, newConfig);
    }, [range, setColorUI]);
    var dropdownSelectProps = {
      disabled: disabled,
      selectedItems: scaleOptions.filter(function (op) {
        return op.value === scaleType;
      }),
      options: scaleOptions,
      multiSelect: false,
      searchable: false,
      onChange: onSelect,
      displayOption: 'label',
      getOptionValue: 'value',
      channelKey: key
    };
    return /*#__PURE__*/_react["default"].createElement(_styledComponents.SidePanelSection, null, /*#__PURE__*/_react["default"].createElement(_styledComponents.PanelLabel, null, /*#__PURE__*/_react["default"].createElement(_localization.FormattedMessage, {
      id: label ? "scale.".concat((0, _utils.camelize)(label)) : 'misc.scale',
      defaultMessage: label
    })), isColorScale && dataset ? /*#__PURE__*/_react["default"].createElement(ColorScaleSelector, (0, _extends2["default"])({}, dropdownSelectProps, {
      layer: layer,
      field: layer.config[field],
      dataset: dataset,
      onSelect: onSelect,
      scaleType: scaleType,
      domain: layer.config[domain],
      aggregatedBins: layer.config.aggregatedBins,
      range: layer.config.visConfig[range],
      setColorUI: _setColorUI,
      colorUIConfig: (_layer$config$colorUI = layer.config.colorUI) === null || _layer$config$colorUI === void 0 ? void 0 : _layer$config$colorUI[range]
    })) : /*#__PURE__*/_react["default"].createElement(SizeScaleSelector, dropdownSelectProps));
  };
  return DimensionScaleSelector;
}
var _default = exports["default"] = DimensionScaleSelectorFactory;
