// kepler.gl/dist/dist/side-panel/layer-panel/vis-config-by-field-selector.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _react = _interopRequireDefault(require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/node_modules/react"));
var _reactIntl = require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/node_modules/react-intl");
var _utils = require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/src/utils/src");
var _fieldSelector = _interopRequireDefault(require("../../common/field-selector"));
var _infoHelper = _interopRequireDefault(require("../../common/info-helper"));
var _styledComponents = require("../../common/styled-components");
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

VisConfigByFieldSelectorFactory.deps = [_infoHelper["default"], _fieldSelector["default"]];
function VisConfigByFieldSelectorFactory(InfoHelper, FieldSelector) {
  var VisConfigByFieldSelector = function VisConfigByFieldSelector(_ref) {
    var id = _ref.id,
      property = _ref.property,
      selectedField = _ref.selectedField,
      description = _ref.description,
      label = _ref.label,
      intl = _ref.intl,
      updateField = _ref.updateField,
      fields = _ref.fields,
      placeholder = _ref.placeholder,
      disabled = _ref.disabled;
    return /*#__PURE__*/_react["default"].createElement(_styledComponents.SidePanelSection, {
      disabled: disabled
    }, /*#__PURE__*/_react["default"].createElement(_styledComponents.PanelLabelWrapper, null, /*#__PURE__*/_react["default"].createElement(_styledComponents.PanelLabel, null, label && /*#__PURE__*/_react["default"].createElement(_reactIntl.FormattedMessage, {
      id: label
    }) || /*#__PURE__*/_react["default"].createElement(_reactIntl.FormattedMessage, {
      id: "layer.propertyBasedOn",
      values: {
        property: intl.formatMessage({
          id: "property.".concat((0, _utils.camelize)(property)),
          defaultMessage: property
        })
      }
    })), description && /*#__PURE__*/_react["default"].createElement(InfoHelper, {
      description: description,
      property: property,
      id: "".concat(id, "-").concat(property)
    })), /*#__PURE__*/_react["default"].createElement(FieldSelector, {
      fields: fields,
      value: selectedField && selectedField.name,
      placeholder: placeholder,
      onSelect: function onSelect(items) {
        return updateField(Array.isArray(items) ? items === null || items === void 0 ? void 0 : items[0] : items);
      },
      disabled: disabled,
      erasable: true
    }));
  };
  return (0, _reactIntl.injectIntl)(VisConfigByFieldSelector);
}
var _default = exports["default"] = VisConfigByFieldSelectorFactory;
