// kepler.gl/src/styles/dist/base.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.inputBgdHover = exports.inputBgdActiveLT = exports.inputBgdActive = exports.inputBgd = exports.histogramOverlayColor = exports.histogramFillOutRange = exports.histogramFillInRange = exports.histogramBreakLineColor = exports.geocoderWidth = exports.geocoderTop = exports.geocoderRight = exports.geocoderInputHeight = exports.fontWhiteColor = exports.fontWeight = exports.fontSize = exports.fontFamily = exports.floatingBtnRadius = exports.floatingBtnColor = exports.floatingBtnBorderHover = exports.floatingBtnBorder = exports.floatingBtnBgdHover = exports.floatingBtnBgd = exports.floatingBtnActColor = exports.floatingBtnActBgd = exports.fieldTokenWidth = exports.fieldTokenRightMargin = exports.fieldTokenHeight = exports.errorColor = exports.effectTypeIconSizeL = exports.effectTypeIconMarginSide = exports.effectTypeIconBgHoverColor = exports.effectPanelWidth = exports.effectPanelTextSecondary3 = exports.effectPanelTextSecondary2 = exports.effectPanelTextSecondary1 = exports.effectPanelTextMain = exports.effectPanelPaddingTop = exports.effectPanelPaddingSide = exports.effectPanelHeight = exports.effectPanelElementColorSun = exports.effectPanelElementColorSelected = exports.effectPanelElementColorHovered = exports.effectPanelElementColorActive = exports.effectPanelElementColor = exports.effectPanelAddEffectFontFamily = exports.effectConfiguratorPadding = exports.effectConfiguratorMargin = exports.dropdownWrapperZ = exports.dropdownWapperMargin = exports.dropdownSelectHeight = exports.dropdownListShadow = exports.dropdownListLineHeight = exports.dropdownListHighlightBgLT = exports.dropdownListHighlightBg = exports.dropdownListBorderTopLT = exports.dropdownListBorderTop = exports.dropdownListBgdLT = exports.dropdownListBgd = exports.dndOverBackgroundColor = exports.dataTableTextColor = exports.customRangeInputWidth = exports.ctaBtnColor = exports.ctaBtnBgdHover = exports.ctaBtnBgd = exports.ctaBtnActColor = exports.ctaBtnActBgd = exports.chickletBgdLT = exports.chickletBgd = exports.checkboxWidth = exports.checkboxMargin = exports.checkboxHeight = exports.checkboxBoxBgdChecked = exports.checkboxBoxBgd = exports.checkboxBorderRadius = exports.checkboxBorderColorLT = exports.checkboxBorderColor = exports.btnFontFamily = exports.breakPoints = exports.boxSizing = exports.boxShadow = exports.bottomWidgetPaddingTop = exports.bottomWidgetPaddingRight = exports.bottomWidgetPaddingLeft = exports.bottomWidgetPaddingBottom = exports.bottomWidgetBgd = exports.bottomPanelGapPalm = exports.bottomPanelGap = exports.bottomInnerPdVert = exports.bottomInnerPdSide = exports.borderRadius = exports.borderColorLT = exports.borderColor = exports.axisFontSize = exports.axisFontColor = exports.aiAssistantPanelWidth = exports.activeColorLT = exports.activeColorHover = exports.activeColor = exports.actionPanelWidth = exports.actionPanelHeight = void 0;
exports.panelToggleBottomPadding = exports.panelToggleBorderColor = exports.panelTabWidth = exports.panelTabColor = exports.panelHeaderIconHover = exports.panelHeaderIconActive = exports.panelHeaderIcon = exports.panelHeaderHeight = exports.panelHeaderBorderRadius = exports.panelContentBackground = exports.panelBoxShadow = exports.panelBorderRadius = exports.panelBorderLT = exports.panelBorderColor = exports.panelBorder = exports.panelBackgroundLT = exports.panelBackgroundHover = exports.panelBackground = exports.notificationPanelWidth = exports.notificationPanelItemWidth = exports.notificationPanelItemHeight = exports.notificationColors = exports.negativeBtnColor = exports.negativeBtnBorder = exports.negativeBtnBgdHover = exports.negativeBtnBgd = exports.negativeBtnActColor = exports.negativeBtnActBgd = exports.modalTitleZ = exports.modalTitleFontSizeSmaller = exports.modalTitleFontSize = exports.modalTitleColor = exports.modalScrollBar = exports.modalPortableLateralPadding = exports.modalPadding = exports.modalOverlayBgd = exports.modalOverLayZ = exports.modalLateralPadding = exports.modalImagePlaceHolder = exports.modalFooterZ = exports.modalFooterBgd = exports.modalDropdownBackground = exports.modalDialogColor = exports.modalDialogBgd = exports.modalContentZ = exports.modalButtonZ = exports.mapPanelHeaderBackgroundColor = exports.mapPanelBackgroundColor = exports.mapControlTop = exports.logoColor = exports.linkBtnColor = exports.linkBtnBorder = exports.linkBtnBgd = exports.linkBtnActColor = exports.linkBtnActBgdHover = exports.linkBtnActBgd = exports.lineHeight = exports.layerTypeIconSizeSM = exports.layerTypeIconSizeL = exports.layerTypeIconPdL = exports.layerPanelToggleOptionColorActive = exports.layerPanelToggleOptionColor = exports.layerPanelHeaderHeight = exports.layerConfiguratorPadding = exports.layerConfiguratorMargin = exports.layerConfiguratorBorderColor = exports.layerConfiguratorBorder = exports.layerConfigGroupPaddingLeft = exports.layerConfigGroupMarginBottom = exports.layerConfigGroupLabelPadding = exports.layerConfigGroupLabelMargin = exports.layerConfigGroupLabelLabelMargin = exports.layerConfigGroupLabelLabelFontSize = exports.layerConfigGroupLabelBorderLeft = exports.layerConfigGroupColor = exports.labelHoverColor = exports.labelColorLT = exports.labelColor = exports.inputPlaceholderFontWeight = exports.inputPlaceholderColorLT = exports.inputPlaceholderColor = exports.inputPaddingTiny = exports.inputPaddingSmall = exports.inputPadding = exports.inputFontWeight = exports.inputFontSizeSmall = exports.inputFontSize = exports.inputColor = exports.inputBoxShadowActiveLT = exports.inputBoxShadowActive = exports.inputBoxShadow = exports.inputBoxHeightTiny = exports.inputBoxHeightSmall = exports.inputBoxHeight = exports.inputBorderRadius = exports.inputBorderHoverColorLT = exports.inputBorderHoverColor = exports.inputBorderColor = exports.inputBorderActiveColorLT = exports.inputBorderActiveColor = void 0;
exports.sliderInputPadding = exports.sliderInputHeight = exports.sliderInputFontSize = exports.sliderInactiveBorderColor = exports.sliderHandleWidth = exports.sliderHandleTextColor = exports.sliderHandleShadow = exports.sliderHandleHoverColor = exports.sliderHandleHeight = exports.sliderHandleColor = exports.sliderHandleAfterContent = exports.sliderBorderRadius = exports.sliderBarRadius = exports.sliderBarHoverColor = exports.sliderBarHeight = exports.sliderBarColor = exports.sliderBarBgd = exports.sidepanelDividerMargin = exports.sidepanelDividerHeight = exports.sidepanelDividerBorder = exports.sidePanelTitleLineHeight = exports.sidePanelTitleFontsize = exports.sidePanelScrollBarWidth = exports.sidePanelScrollBarHeight = exports.sidePanelInnerPadding = exports.sidePanelHeaderBorder = exports.sidePanelHeaderBg = exports.sidePanelBorderColor = exports.sidePanelBorder = exports.sidePanelBg = exports.sideBarCloseBtnColor = exports.sideBarCloseBtnBgdHover = exports.sideBarCloseBtnBgd = exports.selectionBtnColor = exports.selectionBtnBorderColor = exports.selectionBtnBorderActColor = exports.selectionBtnBorder = exports.selectionBtnBgdHover = exports.selectionBtnBgd = exports.selectionBtnActColor = exports.selectionBtnActBgd = exports.selectFontWeightBold = exports.selectFontWeight = exports.selectFontSize = exports.selectColorPlaceHolderLT = exports.selectColorPlaceHolder = exports.selectColorLT = exports.selectColor = exports.selectBorderRadius = exports.selectBorderColorLT = exports.selectBorderColor = exports.selectBorder = exports.selectBackgroundLT = exports.selectBackgroundHoverLT = exports.selectBackgroundHover = exports.selectBackground = exports.selectActiveBorderColor = exports.secondarySwitchTrackBgd = exports.secondarySwitchBtnBgd = exports.secondaryInputColor = exports.secondaryInputBorderColor = exports.secondaryInputBorderActiveColor = exports.secondaryInputBgdHover = exports.secondaryInputBgdActive = exports.secondaryInputBgd = exports.secondaryBtnColor = exports.secondaryBtnBorder = exports.secondaryBtnBgdHover = exports.secondaryBtnBgd = exports.secondaryBtnActColor = exports.secondaryBtnActBgd = exports.scrollbarThumbColorLT = exports.scrollbarThumbColorHoverLT = exports.rightPanelMarginTop = exports.rightPanelMarginRight = exports.rangePlotMarginLarge = exports.rangePlotMargin = exports.rangePlotHLargePalm = exports.rangePlotHLarge = exports.rangePlotH = exports.rangePlotContainerHLargePalm = exports.rangePlotContainerHLarge = exports.rangePlotContainerH = exports.rangeBrushBgd = exports.radioRadius = exports.radioButtonRadius = exports.radioButtonBgdColor = exports.radioBorderRadius = exports.radioBorderColor = exports.primaryBtnRadius = exports.primaryBtnFontSizeSmall = exports.primaryBtnFontSizeLarge = exports.primaryBtnFontSizeDefault = exports.primaryBtnColor = exports.primaryBtnBorder = exports.primaryBtnBgdHover = exports.primaryBtnBgd = exports.primaryBtnActColor = exports.primaryBtnActBgd = exports.panelToggleMarginRight = void 0;
exports.transitionSlow = exports.transitionFast = exports.transition = exports.tooltipFontSize = exports.tooltipColorLT = exports.tooltipColor = exports.tooltipBoxShadow = exports.tooltipBgLT = exports.tooltipBg = exports.toolbarItemIconHover = exports.toolbarItemBorderRaddius = exports.toolbarItemBorderHover = exports.toolbarItemBgdHover = exports.titleTextColor = exports.titleColorLT = exports.timeTitleFontSize = exports.themeLT = exports.themeBS = exports.theme = exports.textTruncate = exports.textColorLT = exports.textColorHlLT = exports.textColorHl = exports.textColor = exports.switchWidth = exports.switchTrackBorderRadius = exports.switchTrackBgdActive = exports.switchTrackBgd = exports.switchLabelMargin = exports.switchHeight = exports.switchBtnWidth = exports.switchBtnHeight = exports.switchBtnBoxShadow = exports.switchBtnBorderRadius = exports.switchBtnBgdActive = exports.switchBtnBgd = exports.subtextColorLT = exports.subtextColorActive = exports.subtextColor = exports.styledConfigGroupHeaderBorder = exports.sliderMarginTopIsTime = exports.sliderMarginTop = exports.sliderMarginBottom = exports.sliderInputWidth = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _taggedTemplateLiteral2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteral"));
var _styledComponents = require("styled-components");
var _constants = require("@kepler.gl/constants");
var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5, _templateObject6, _templateObject7, _templateObject8, _templateObject9, _templateObject10, _templateObject11, _templateObject12, _templateObject13, _templateObject14, _templateObject15, _templateObject16, _templateObject17, _templateObject18, _templateObject19, _templateObject20, _templateObject21, _templateObject22, _templateObject23, _templateObject24, _templateObject25, _templateObject26, _templateObject27, _templateObject28, _templateObject29, _templateObject30, _templateObject31, _templateObject32, _templateObject33, _templateObject34; // SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2["default"])(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var transition = exports.transition = 'all .4s ease';
var transitionFast = exports.transitionFast = 'all .2s ease';
var transitionSlow = exports.transitionSlow = 'all .8s ease';
var boxShadow = exports.boxShadow = '0 1px 2px 0 rgba(0,0,0,0.10)';
var boxSizing = exports.boxSizing = 'border-box';
var borderRadius = exports.borderRadius = '1px';
var borderColor = exports.borderColor = '#3A414C';
var borderColorLT = exports.borderColorLT = '#F1F1F1';

// TEXT
var fontFamily = exports.fontFamily = "ff-clan-web-pro, 'Helvetica Neue', Helvetica, sans-serif";
var fontWeight = exports.fontWeight = 400;
var fontSize = exports.fontSize = '0.875em';
var lineHeight = exports.lineHeight = 1.71429;
var labelColor = exports.labelColor = '#6A7485';
var labelHoverColor = exports.labelHoverColor = '#C6C6C6';
var labelColorLT = exports.labelColorLT = '#6A7485';
var textColor = exports.textColor = '#A0A7B4';
var textColorLT = exports.textColorLT = '#3A414C';
var dataTableTextColor = exports.dataTableTextColor = textColorLT;
var titleColorLT = exports.titleColorLT = '#29323C';
var subtextColor = exports.subtextColor = '#6A7485';
var subtextColorLT = exports.subtextColorLT = '#A0A7B4';
var subtextColorActive = exports.subtextColorActive = '#FFFFFF';
var fontWhiteColor = exports.fontWhiteColor = '#54638c';
var panelToggleBorderColor = exports.panelToggleBorderColor = '#FFFFFF';
var panelTabWidth = exports.panelTabWidth = '30px';
var titleTextColor = exports.titleTextColor = '#FFFFFF';
var textColorHl = exports.textColorHl = '#F0F0F0';
var textColorHlLT = exports.textColorHlLT = '#000000';
var activeColor = exports.activeColor = '#1FBAD6';
var activeColorLT = exports.activeColorLT = '#2473BD';
var activeColorHover = exports.activeColorHover = '#108188';
var errorColor = exports.errorColor = '#F9042C';
var logoColor = exports.logoColor = activeColor;

// Button
var btnFontFamily = exports.btnFontFamily = fontFamily;
var primaryBtnBgd = exports.primaryBtnBgd = '#0F9668';
var primaryBtnActBgd = exports.primaryBtnActBgd = '#13B17B';
var primaryBtnColor = exports.primaryBtnColor = '#FFFFFF';
var primaryBtnActColor = exports.primaryBtnActColor = '#FFFFFF';
var primaryBtnBgdHover = exports.primaryBtnBgdHover = '#13B17B';
var primaryBtnRadius = exports.primaryBtnRadius = '2px';
var primaryBtnFontSizeDefault = exports.primaryBtnFontSizeDefault = '11px';
var primaryBtnFontSizeSmall = exports.primaryBtnFontSizeSmall = '10px';
var primaryBtnFontSizeLarge = exports.primaryBtnFontSizeLarge = '14px';
var primaryBtnBorder = exports.primaryBtnBorder = '0';
var secondaryBtnBgd = exports.secondaryBtnBgd = '#6A7485';
var secondaryBtnActBgd = exports.secondaryBtnActBgd = '#A0A7B4';
var secondaryBtnColor = exports.secondaryBtnColor = '#FFFFFF';
var secondaryBtnActColor = exports.secondaryBtnActColor = '#FFFFFF';
var secondaryBtnBgdHover = exports.secondaryBtnBgdHover = '#A0A7B4';
var secondaryBtnBorder = exports.secondaryBtnBorder = '0';
var ctaBtnBgd = exports.ctaBtnBgd = '#0F9668';
var ctaBtnBgdHover = exports.ctaBtnBgdHover = '#13B17B';
var ctaBtnActBgd = exports.ctaBtnActBgd = '#13B17B';
var ctaBtnColor = exports.ctaBtnColor = '#FFFFFF';
var ctaBtnActColor = exports.ctaBtnActColor = '#FFFFFF';
var linkBtnBgd = exports.linkBtnBgd = 'transparent';
var linkBtnActBgd = exports.linkBtnActBgd = linkBtnBgd;
var linkBtnColor = exports.linkBtnColor = '#A0A7B4';
var linkBtnActColor = exports.linkBtnActColor = '#F1F1F1';
var linkBtnActBgdHover = exports.linkBtnActBgdHover = linkBtnBgd;
var linkBtnBorder = exports.linkBtnBorder = '0';
var negativeBtnBgd = exports.negativeBtnBgd = errorColor;
var negativeBtnActBgd = exports.negativeBtnActBgd = '#FF193E';
var negativeBtnBgdHover = exports.negativeBtnBgdHover = '#FF193E';
var negativeBtnBorder = exports.negativeBtnBorder = '0';
var negativeBtnColor = exports.negativeBtnColor = '#FFFFFF';
var negativeBtnActColor = exports.negativeBtnActColor = '#FFFFFF';
var floatingBtnBgd = exports.floatingBtnBgd = '#29323C';
var floatingBtnActBgd = exports.floatingBtnActBgd = '#3A4552';
var floatingBtnBgdHover = exports.floatingBtnBgdHover = '#3A4552';
var floatingBtnBorder = exports.floatingBtnBorder = '0';
var floatingBtnBorderHover = exports.floatingBtnBorderHover = '0';
var floatingBtnColor = exports.floatingBtnColor = subtextColor;
var floatingBtnActColor = exports.floatingBtnActColor = subtextColorActive;
var floatingBtnRadius = exports.floatingBtnRadius = '0px';
var selectionBtnBgd = exports.selectionBtnBgd = 'transparent';
var selectionBtnActBgd = exports.selectionBtnActBgd = 'transparent';
var selectionBtnColor = exports.selectionBtnColor = '#D3D8E0';
var selectionBtnActColor = exports.selectionBtnActColor = '#0F9668';
var selectionBtnBgdHover = exports.selectionBtnBgdHover = '#0F9668';
var selectionBtnBorder = exports.selectionBtnBorder = '1';
var selectionBtnBorderColor = exports.selectionBtnBorderColor = '#D3D8E0';
var selectionBtnBorderActColor = exports.selectionBtnBorderActColor = '#0F9668';
// Scrollbar
var scrollbarThumbColorLT = exports.scrollbarThumbColorLT = labelColorLT;
var scrollbarThumbColorHoverLT = exports.scrollbarThumbColorHoverLT = textColorHlLT;
// Input
var inputBoxHeight = exports.inputBoxHeight = '34px';
var inputBoxHeightSmall = exports.inputBoxHeightSmall = '24px';
var inputBoxHeightTiny = exports.inputBoxHeightTiny = '18px';
var inputPadding = exports.inputPadding = '4px 10px';
var inputPaddingSmall = exports.inputPaddingSmall = '4px 6px';
var inputPaddingTiny = exports.inputPaddingTiny = '2px 4px';
var inputFontSize = exports.inputFontSize = '11px';
var inputFontSizeSmall = exports.inputFontSizeSmall = '10px';
var inputFontWeight = exports.inputFontWeight = 500;
var inputBgd = exports.inputBgd = '#29323C';
var inputBgdHover = exports.inputBgdHover = '#3A414C';
var inputBgdActive = exports.inputBgdActive = '#3A414C';
var inputBgdActiveLT = exports.inputBgdActiveLT = '#FFFFFF';
var inputBorderColor = exports.inputBorderColor = '#29323C';
var inputBorderHoverColor = exports.inputBorderHoverColor = '#3A414C';
var inputBorderHoverColorLT = exports.inputBorderHoverColorLT = subtextColor;
var inputBorderActiveColor = exports.inputBorderActiveColor = '#3A414C';
var inputBorderActiveColorLT = exports.inputBorderActiveColorLT = textColorLT;
var inputColor = exports.inputColor = '#A0A7B4';
var inputBorderRadius = exports.inputBorderRadius = '1px';
var inputPlaceholderColor = exports.inputPlaceholderColor = '#6A7485';
var inputPlaceholderColorLT = exports.inputPlaceholderColorLT = subtextColorLT;
var inputPlaceholderFontWeight = exports.inputPlaceholderFontWeight = 400;
var inputBoxShadow = exports.inputBoxShadow = 'none';
var inputBoxShadowActive = exports.inputBoxShadowActive = 'none';
var inputBoxShadowActiveLT = exports.inputBoxShadowActiveLT = 'none';
var secondaryInputBgd = exports.secondaryInputBgd = '#242730';
var secondaryInputBgdHover = exports.secondaryInputBgdHover = '#3A414C';
var secondaryInputBgdActive = exports.secondaryInputBgdActive = '#3A414C';
var secondaryInputColor = exports.secondaryInputColor = '#A0A7B4';
var secondaryInputBorderColor = exports.secondaryInputBorderColor = '#242730';
var secondaryInputBorderActiveColor = exports.secondaryInputBorderActiveColor = '#D3D8E0';
var dropdownSelectHeight = exports.dropdownSelectHeight = 30;

// Select
var selectColor = exports.selectColor = inputColor;
var selectColorLT = exports.selectColorLT = titleColorLT;
var selectActiveBorderColor = exports.selectActiveBorderColor = '#D3D8E0';
var selectFontSize = exports.selectFontSize = '11px';
var selectFontWeight = exports.selectFontWeight = '400';
var selectFontWeightBold = exports.selectFontWeightBold = '500';
var selectColorPlaceHolder = exports.selectColorPlaceHolder = '#6A7485';
var selectColorPlaceHolderLT = exports.selectColorPlaceHolderLT = selectColorLT;
var selectBackground = exports.selectBackground = inputBgd;
var selectBackgroundHover = exports.selectBackgroundHover = inputBgdHover;
var selectBackgroundLT = exports.selectBackgroundLT = '#FFFFFF';
var selectBackgroundHoverLT = exports.selectBackgroundHoverLT = '#F8F8F9';
var selectBorderColor = exports.selectBorderColor = '#D3D8E0';
var selectBorderColorLT = exports.selectBorderColorLT = '#D3D8E0';
var selectBorderRadius = exports.selectBorderRadius = '1px';
var selectBorder = exports.selectBorder = 0;
var panelTabColor = exports.panelTabColor = subtextColor;
var dropdownListHighlightBg = exports.dropdownListHighlightBg = '#6A7485';
var dropdownListHighlightBgLT = exports.dropdownListHighlightBgLT = '#F8F8F9';
var dropdownListShadow = exports.dropdownListShadow = '0 6px 12px 0 rgba(0,0,0,0.16)';
var dropdownListBgd = exports.dropdownListBgd = '#29323C';
var toolbarItemBgdHover = exports.toolbarItemBgdHover = '#3A4552';
var toolbarItemIconHover = exports.toolbarItemIconHover = textColorHl;
var toolbarItemBorderHover = exports.toolbarItemBorderHover = 'transparent';
var toolbarItemBorderRaddius = exports.toolbarItemBorderRaddius = '0px';
var dropdownListBgdLT = exports.dropdownListBgdLT = '#FFFFFF';
var dropdownListBorderTop = exports.dropdownListBorderTop = '#242730';
var dropdownListBorderTopLT = exports.dropdownListBorderTopLT = '#D3D8E0';
var dropdownListLineHeight = exports.dropdownListLineHeight = 20;
var dropdownWrapperZ = exports.dropdownWrapperZ = 100;
var dropdownWapperMargin = exports.dropdownWapperMargin = 4;
var dndOverBackgroundColor = exports.dndOverBackgroundColor = 'rgba(128, 128, 128, 0.2)';

// Switch
var switchWidth = exports.switchWidth = 24;
var switchHeight = exports.switchHeight = 12;
var switchLabelMargin = exports.switchLabelMargin = 12;
var switchTrackBgd = exports.switchTrackBgd = '#29323C';
var switchTrackBgdActive = exports.switchTrackBgdActive = activeColor;
var switchTrackBorderRadius = exports.switchTrackBorderRadius = '1px';
var switchBtnBgd = exports.switchBtnBgd = '#6A7485';
var switchBtnBgdActive = exports.switchBtnBgdActive = '#D3D8E0';
var switchBtnBoxShadow = exports.switchBtnBoxShadow = '0 2px 4px 0 rgba(0,0,0,0.40)';
var switchBtnBorderRadius = exports.switchBtnBorderRadius = '0';
var switchBtnWidth = exports.switchBtnWidth = 12;
var switchBtnHeight = exports.switchBtnHeight = 12;
var secondarySwitchTrackBgd = exports.secondarySwitchTrackBgd = '#242730';
var secondarySwitchBtnBgd = exports.secondarySwitchBtnBgd = '#3A414C';

// Checkbox
var checkboxWidth = exports.checkboxWidth = 16;
var checkboxHeight = exports.checkboxHeight = 16;
var checkboxMargin = exports.checkboxMargin = 12;
var checkboxBorderColor = exports.checkboxBorderColor = selectBorderColor;
var checkboxBorderRadius = exports.checkboxBorderRadius = '2px';
var checkboxBorderColorLT = exports.checkboxBorderColorLT = selectBorderColorLT;
var checkboxBoxBgd = exports.checkboxBoxBgd = 'white';
var checkboxBoxBgdChecked = exports.checkboxBoxBgdChecked = primaryBtnBgd;

// Radio
var radioRadius = exports.radioRadius = 8;
var radioBorderRadius = exports.radioBorderRadius = 100;
var radioBorderColor = exports.radioBorderColor = 'transparent';
var radioButtonRadius = exports.radioButtonRadius = 4;
var radioButtonBgdColor = exports.radioButtonBgdColor = switchBtnBgdActive;

// Side Panel
var sidePanelHeaderBg = exports.sidePanelHeaderBg = '#29323C';
var sidePanelHeaderBorder = exports.sidePanelHeaderBorder = 'transparent';
var layerConfigGroupMarginBottom = exports.layerConfigGroupMarginBottom = 12;
var layerConfigGroupPaddingLeft = exports.layerConfigGroupPaddingLeft = 18;
var sidePanelInnerPadding = exports.sidePanelInnerPadding = 16;
var sidePanelBorder = exports.sidePanelBorder = 0;
var sidePanelBorderColor = exports.sidePanelBorderColor = 'transparent';
var sidePanelBg = exports.sidePanelBg = '#242730';
var sidePanelScrollBarWidth = exports.sidePanelScrollBarWidth = 10;
var sidePanelScrollBarHeight = exports.sidePanelScrollBarHeight = 10;
var sideBarCloseBtnBgd = exports.sideBarCloseBtnBgd = secondaryBtnBgd;
var sideBarCloseBtnColor = exports.sideBarCloseBtnColor = '#29323C';
var sideBarCloseBtnBgdHover = exports.sideBarCloseBtnBgdHover = secondaryBtnActBgd;
var sidePanelTitleFontsize = exports.sidePanelTitleFontsize = '20px';
var sidePanelTitleLineHeight = exports.sidePanelTitleLineHeight = '1.71429';
var panelBackground = exports.panelBackground = '#29323C';
var panelContentBackground = exports.panelContentBackground = '#292E36';
var panelBackgroundHover = exports.panelBackgroundHover = '#3A4552';
var panelHeaderBorderRadius = exports.panelHeaderBorderRadius = '0px';
var chickletBgd = exports.chickletBgd = '#3A4552';
var chickletBgdLT = exports.chickletBgdLT = '#D3D8E0';
var panelHeaderIcon = exports.panelHeaderIcon = '#6A7485';
var panelHeaderIconActive = exports.panelHeaderIconActive = '#A0A7B4';
var panelHeaderIconHover = exports.panelHeaderIconHover = textColorHl;
var panelHeaderHeight = exports.panelHeaderHeight = 48;
var layerPanelHeaderHeight = exports.layerPanelHeaderHeight = 48;
var panelBoxShadow = exports.panelBoxShadow = '0 6px 12px 0 rgba(0,0,0,0.16)';
var panelBorderRadius = exports.panelBorderRadius = '2px';
var panelBackgroundLT = exports.panelBackgroundLT = '#F8F8F9';
var panelToggleMarginRight = exports.panelToggleMarginRight = 12;
var panelToggleBottomPadding = exports.panelToggleBottomPadding = 6;
var panelBorderColor = exports.panelBorderColor = '#3A414C';
var panelBorder = exports.panelBorder = "1px solid ".concat(borderColor);
var panelBorderLT = exports.panelBorderLT = "1px solid ".concat(borderColorLT);
var mapPanelBackgroundColor = exports.mapPanelBackgroundColor = '#242730';
var mapPanelHeaderBackgroundColor = exports.mapPanelHeaderBackgroundColor = '#29323C';
var tooltipBg = exports.tooltipBg = '#3A414C';
var tooltipBgLT = exports.tooltipBgLT = '#1869B5';
var tooltipColor = exports.tooltipColor = textColorHl;
var tooltipColorLT = exports.tooltipColorLT = '#FFFFFF';
var tooltipBoxShadow = exports.tooltipBoxShadow = boxShadow;
var tooltipFontSize = exports.tooltipFontSize = '10px';
var layerTypeIconSizeL = exports.layerTypeIconSizeL = 50;
var layerTypeIconPdL = exports.layerTypeIconPdL = 12;
var layerTypeIconSizeSM = exports.layerTypeIconSizeSM = 28;
var layerPanelToggleOptionColor = exports.layerPanelToggleOptionColor = '#6A7485';
var layerPanelToggleOptionColorActive = exports.layerPanelToggleOptionColorActive = '#F0F0F0';

// Sidepanel divider
var sidepanelDividerBorder = exports.sidepanelDividerBorder = '1px';
var sidepanelDividerMargin = exports.sidepanelDividerMargin = 12;
var sidepanelDividerHeight = exports.sidepanelDividerHeight = 12;

// Bottom Panel
var bottomInnerPdSide = exports.bottomInnerPdSide = 32;
var bottomInnerPdVert = exports.bottomInnerPdVert = 6;
var bottomPanelGap = exports.bottomPanelGap = 20;
var bottomPanelGapPalm = exports.bottomPanelGapPalm = 20;
var bottomWidgetPaddingTop = exports.bottomWidgetPaddingTop = 20;
var bottomWidgetPaddingRight = exports.bottomWidgetPaddingRight = 20;
var bottomWidgetPaddingBottom = exports.bottomWidgetPaddingBottom = 30;
var bottomWidgetPaddingLeft = exports.bottomWidgetPaddingLeft = 20;
var bottomWidgetBgd = exports.bottomWidgetBgd = '#29323C';
// Modal
var modalTitleColor = exports.modalTitleColor = '#3A414C';
var modalTitleFontSize = exports.modalTitleFontSize = '24px';
var modalTitleFontSizeSmaller = exports.modalTitleFontSizeSmaller = '18px';
var modalFooterBgd = exports.modalFooterBgd = '#F8F8F9';
var modalImagePlaceHolder = exports.modalImagePlaceHolder = '#DDDFE3';
var modalPadding = exports.modalPadding = '10px 0';
var modalLateralPadding = exports.modalLateralPadding = '72px';
var modalPortableLateralPadding = exports.modalPortableLateralPadding = '36px';
var modalOverLayZ = exports.modalOverLayZ = 1001;
var modalOverlayBgd = exports.modalOverlayBgd = 'rgba(0, 0, 0, 0.5)';
var modalContentZ = exports.modalContentZ = 10002;
var modalFooterZ = exports.modalFooterZ = 10001;
var modalTitleZ = exports.modalTitleZ = 10003;
var modalButtonZ = exports.modalButtonZ = 10005;
var modalDropdownBackground = exports.modalDropdownBackground = '#FFFFFF';

// Modal Dialog (Dark)
var modalDialogBgd = exports.modalDialogBgd = '#3A414C';
var modalDialogColor = exports.modalDialogColor = textColorHl;

// Slider
var sliderBarColor = exports.sliderBarColor = '#6A7485';
var sliderBarBgd = exports.sliderBarBgd = '#3A414C';
var sliderBarHoverColor = exports.sliderBarHoverColor = '#D3D8E0';
var sliderBarRadius = exports.sliderBarRadius = '1px';
var sliderBarHeight = exports.sliderBarHeight = 4;
var sliderHandleHeight = exports.sliderHandleHeight = 12;
var sliderHandleWidth = exports.sliderHandleWidth = 12;
var sliderHandleColor = exports.sliderHandleColor = '#D3D8E0';
var sliderHandleTextColor = exports.sliderHandleTextColor = sliderHandleColor;
var sliderInactiveBorderColor = exports.sliderInactiveBorderColor = sliderHandleColor;
var sliderBorderRadius = exports.sliderBorderRadius = '0';
var sliderHandleHoverColor = exports.sliderHandleHoverColor = '#FFFFFF';
var sliderHandleAfterContent = exports.sliderHandleAfterContent = '';
var sliderHandleShadow = exports.sliderHandleShadow = '0 2px 4px 0 rgba(0,0,0,0.40)';
var sliderInputHeight = exports.sliderInputHeight = 24;
var sliderInputWidth = exports.sliderInputWidth = 56;
var sliderInputFontSize = exports.sliderInputFontSize = '10px';
var sliderInputPadding = exports.sliderInputPadding = '4px 6px';
var sliderMarginTopIsTime = exports.sliderMarginTopIsTime = -12;
var sliderMarginTop = exports.sliderMarginTop = 12;
var sliderMarginBottom = exports.sliderMarginBottom = 12;

// Custom Range Input
var customRangeInputWidth = exports.customRangeInputWidth = 100;

// Geocoder
var geocoderWidth = exports.geocoderWidth = 360;
var geocoderTop = exports.geocoderTop = 20;
var geocoderRight = exports.geocoderRight = 12;
var geocoderInputHeight = exports.geocoderInputHeight = 36;

// Map Control
var mapControlTop = exports.mapControlTop = 52;

// Plot
var rangeBrushBgd = exports.rangeBrushBgd = '#3A414C';
var histogramFillInRange = exports.histogramFillInRange = activeColor;
var histogramFillOutRange = exports.histogramFillOutRange = sliderBarColor;
var histogramOverlayColor = exports.histogramOverlayColor = '#999999';
var histogramBreakLineColor = exports.histogramBreakLineColor = '#505b7c';
var axisFontSize = exports.axisFontSize = '10px';
var axisFontColor = exports.axisFontColor = textColor;
var timeTitleFontSize = exports.timeTitleFontSize = '10px';
var rangePlotMargin = exports.rangePlotMargin = {
  top: 12,
  bottom: 0,
  left: 0,
  right: 0
};
var rangePlotMarginLarge = exports.rangePlotMarginLarge = {
  top: 18,
  bottom: 0,
  left: 0,
  right: 0
};
var rangePlotH = exports.rangePlotH = 62;
var rangePlotContainerH = exports.rangePlotContainerH = 78;
var rangePlotHLarge = exports.rangePlotHLarge = 102;
var rangePlotHLargePalm = exports.rangePlotHLargePalm = 102;
var rangePlotContainerHLarge = exports.rangePlotContainerHLarge = 120;
var rangePlotContainerHLargePalm = exports.rangePlotContainerHLargePalm = 120;

// Notification
var notificationColors = exports.notificationColors = {
  info: '#276ef1',
  error: '#f25138',
  success: '#47b275',
  warning: '#ffc043'
};
var notificationPanelWidth = exports.notificationPanelWidth = 240;
var notificationPanelItemWidth = exports.notificationPanelItemWidth = notificationPanelWidth - 60;
var notificationPanelItemHeight = exports.notificationPanelItemHeight = 60;

// Data Table
var headerRowHeight = 70;
var headerStatsControlHeight = 32;
var headerRowWStatsHeight = 364;
var rowHeight = 32;
var headerPaddingTop = 6;
var headerPaddingBottom = 8;
var cellPaddingSide = 10;
var edgeCellPaddingSide = 10;
var cellFontSize = 10;
var gridPaddingSide = 24;
var headerCellBackground = '#FFFFFF';
var headerCellBorderColor = '#E0E0E0';
var headerCellStatsBackground = '#F8FAFF';
var headerCellStatsControlBackground = '#EAF0FF';
var headerCellIconColor = '#666666';
var cellBorderColor = '#E0E0E0';
var evenRowBackground = '#FFFFFF';
var oddRowBackground = '#F7F7F7';
var optionButtonColor = '#6A7485';
var pinnedGridBorderColor = '#E0E0E0';

// Floating Time display
var timeDisplayBorderRadius = 32;
var timeDisplayHeight = 64;
var timeDisplayMinWidth = 176;
var timeDisplayOpacity = 0.8;
var timeDisplayPadding = '0 24px';

// Export map modal
var exportIntraSectionMargin = '8';

// progress bar
var progressBarColor = primaryBtnBgd;
var progressBarTrackColor = '#E8E8E8';
// Action Panel
var actionPanelWidth = exports.actionPanelWidth = 110;
var actionPanelHeight = exports.actionPanelHeight = 32;

// Styled Token
var fieldTokenRightMargin = exports.fieldTokenRightMargin = 4;
var fieldTokenHeight = exports.fieldTokenHeight = 20;
var fieldTokenWidth = exports.fieldTokenWidth = 40;
var textTruncate = exports.textTruncate = {
  maxWidth: '100%',
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap',
  wordWrap: 'normal'
};

// layerConfigGroupLabel
var layerConfigGroupLabelBorderLeft = exports.layerConfigGroupLabelBorderLeft = '2px';
var layerConfigGroupLabelMargin = exports.layerConfigGroupLabelMargin = '-12px';
var layerConfigGroupLabelPadding = exports.layerConfigGroupLabelPadding = '10px';
var layerConfigGroupColor = exports.layerConfigGroupColor = 'transparent';

// layerConfigGroupLabel label
var layerConfigGroupLabelLabelMargin = exports.layerConfigGroupLabelLabelMargin = '0';
var layerConfigGroupLabelLabelFontSize = exports.layerConfigGroupLabelLabelFontSize = '12px';

// styledConfigGroupHeader
var styledConfigGroupHeaderBorder = exports.styledConfigGroupHeaderBorder = '2px';

// layerConfigurator

var layerConfiguratorBorder = exports.layerConfiguratorBorder = '0';
var layerConfiguratorBorderColor = exports.layerConfiguratorBorderColor = '';
var layerConfiguratorMargin = exports.layerConfiguratorMargin = '12px';
var layerConfiguratorPadding = exports.layerConfiguratorPadding = '12px 0 8px 0';
// This breakpoints are used for responsive design
var breakPoints = exports.breakPoints = {
  palm: 588,
  desk: 768
};
var aiAssistantPanelWidth = exports.aiAssistantPanelWidth = 415;

// effect manager
var effectConfiguratorMargin = exports.effectConfiguratorMargin = '18px 0 18px 0';
var effectConfiguratorPadding = exports.effectConfiguratorPadding = '0 0 0 18px';
var effectPanelWidth = exports.effectPanelWidth = 345;
var effectPanelHeight = exports.effectPanelHeight = 180;
var effectPanelPaddingSide = exports.effectPanelPaddingSide = 16;
var effectPanelPaddingTop = exports.effectPanelPaddingTop = 16;
var effectPanelAddEffectFontFamily = exports.effectPanelAddEffectFontFamily = btnFontFamily;
var effectTypeIconMarginSide = exports.effectTypeIconMarginSide = 6;
var effectTypeIconSizeL = exports.effectTypeIconSizeL = 56;
var effectTypeIconBgHoverColor = exports.effectTypeIconBgHoverColor = '#262D40';
var effectPanelTextMain = exports.effectPanelTextMain = '#F7F7F7';
var effectPanelTextSecondary1 = exports.effectPanelTextSecondary1 = '#A0A7B4';
var effectPanelTextSecondary2 = exports.effectPanelTextSecondary2 = '#6A7485';
var effectPanelTextSecondary3 = exports.effectPanelTextSecondary3 = '#5A6475';
var effectPanelElementColor = exports.effectPanelElementColor = inputBgd;
var effectPanelElementColorActive = exports.effectPanelElementColorActive = '#323843';
var effectPanelElementColorHovered = exports.effectPanelElementColorHovered = inputBgdHover;
var effectPanelElementColorSelected = exports.effectPanelElementColorSelected = '#454e5d';
var effectPanelElementColorSun = exports.effectPanelElementColorSun = '#F7B26B';

// right panel
var rightPanelMarginTop = exports.rightPanelMarginTop = 12;
var rightPanelMarginRight = exports.rightPanelMarginRight = 12;

// theme is passed to kepler.gl when it's mounted,
// it is used by styled-components to pass along to
// all child components

var input = (0, _styledComponents.css)(_templateObject || (_templateObject = (0, _taggedTemplateLiteral2["default"])(["\n  &::placeholder {\n    color: ", ";\n    font-weight: ", ";\n  }\n\n  /* Disable Arrows on Number Inputs */\n  &::-webkit-inner-spin-button,\n  &::-webkit-outer-spin-button {\n    -webkit-appearance: none;\n    margin: 0;\n  }\n\n  align-items: center;\n  background-color: ", ";\n  border: 1px solid\n    ", ";\n  border-radius: 2px;\n  caret-color: ", ";\n  color: ", ";\n  display: flex;\n  font-size: ", ";\n  font-weight: ", ";\n  font-family: ", ";\n  height: ", ";\n  justify-content: space-between;\n  outline: none;\n  overflow: hidden;\n  padding: ", ";\n  text-overflow: ellipsis;\n  transition: ", ";\n  white-space: nowrap;\n  width: 100%;\n  word-wrap: normal;\n  pointer-events: ", ";\n  opacity: ", ";\n  box-shadow: ", ";\n\n  &:hover {\n    cursor: ", ";\n    background-color: ", ";\n    border-color: ", ";\n  }\n\n  &:active,\n  &:focus,\n  &.focus,\n  &.active {\n    background-color: ", ";\n    border-color: ", ";\n    box-shadow: ", ";\n  }\n"])), function (props) {
  return props.theme.inputPlaceholderColor;
}, function (props) {
  return props.theme.inputPlaceholderFontWeight;
}, function (props) {
  return props.theme.inputBgd;
}, function (props) {
  return props.active ? props.theme.inputBorderActiveColor : props.error ? props.theme.errorColor : props.theme.inputBgd;
}, function (props) {
  return props.theme.inputBorderActiveColor;
}, function (props) {
  return props.theme.inputColor;
}, function (props) {
  return ['small', 'tiny'].includes(props.size) ? props.theme.inputFontSizeSmall : props.theme.inputFontSize;
}, function (props) {
  return props.theme.inputFontWeight;
}, function (props) {
  return props.theme.fontFamily;
}, function (props) {
  return props.size === 'small' ? props.theme.inputBoxHeightSmall : props.size === 'tiny' ? props.theme.inputBoxHeightTiny : props.theme.inputBoxHeight;
}, function (props) {
  return props.size === 'small' ? props.theme.inputPaddingSmall : props.size === 'tiny' ? props.theme.inputPaddingTiny : props.theme.inputPadding;
}, function (props) {
  return props.theme.transition;
}, function (props) {
  return props.disabled ? 'none' : 'all';
}, function (props) {
  return props.disabled ? 0.5 : 1;
}, function (props) {
  return props.theme.inputBoxShadow;
}, function (props) {
  return props.type === 'number' || props.type === 'text' ? 'text' : 'pointer';
}, function (props) {
  return props.active ? props.theme.inputBgdActive : props.theme.inputBgdHover;
}, function (props) {
  return props.active ? props.theme.inputBorderActiveColor : props.theme.inputBorderHoverColor;
}, function (props) {
  return props.theme.inputBgdActive;
}, function (props) {
  return props.theme.inputBorderActiveColor;
}, function (props) {
  return props.theme.inputBoxShadowActive;
});
var inputLT = (0, _styledComponents.css)(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteral2["default"])(["\n  &::placeholder {\n    color: ", ";\n    font-weight: 400;\n  }\n  ", "\n\n  background-color: ", ";\n  border: 1px solid\n    ", ";\n  color: ", ";\n  caret-color: ", ";\n\n  &:hover {\n    background-color: ", ";\n    cursor: ", ";\n    border-color: ", ";\n  }\n\n  &:active,\n  &:focus,\n  &.focus,\n  &.active {\n    background-color: ", ";\n    border-color: ", ";\n    box-shadow: ", ";\n  }\n"])), function (props) {
  return props.theme.inputPlaceholderColorLT;
}, input, function (props) {
  return props.theme.selectBackgroundLT;
}, function (props) {
  return props.active ? props.theme.selectActiveBorderColor : props.error ? props.theme.errorColor : props.theme.selectBorderColorLT;
}, function (props) {
  return props.theme.selectColorLT;
}, function (props) {
  return props.theme.inputBorderActiveColorLT;
}, function (props) {
  return props.theme.inputBgdActiveLT;
}, function (props) {
  return ['number', 'text'].includes(props.type) ? 'text' : 'pointer';
}, function (props) {
  return props.active ? props.theme.inputBorderActiveColorLT : props.theme.inputBorderHoverColorLT;
}, function (props) {
  return props.theme.inputBgdActiveLT;
}, function (props) {
  return props.theme.inputBorderActiveColorLT;
}, function (props) {
  return props.theme.inputBoxShadowActiveLT;
});
var secondaryInput = (0, _styledComponents.css)(_templateObject3 || (_templateObject3 = (0, _taggedTemplateLiteral2["default"])(["\n  ", "\n  color: ", ";\n  background-color: ", ";\n  border: 1px solid\n    ", ";\n\n  &:hover {\n    cursor: pointer;\n    background-color: ", ";\n    border-color: ", ";\n  }\n\n  &:active,\n  &:focus,\n  &.focus,\n  &.active {\n    background-color: ", ";\n    border-color: ", ";\n  }\n"])), function (props) {
  return props.theme.input;
}, function (props) {
  return props.theme.secondaryInputColor;
}, function (props) {
  return props.theme.secondaryInputBgd;
}, function (props) {
  return props.error ? props.theme.errorColor : props.theme.secondaryInputBorderColor;
}, function (props) {
  return props.theme.secondaryInputBgdHover;
}, function (props) {
  return props.theme.secondaryInputBgdHover;
}, function (props) {
  return props.theme.secondaryInputBgdActive;
}, function (props) {
  return props.theme.secondaryInputBorderActiveColor;
});
var chickletedInputContainer = (0, _styledComponents.css)(_templateObject4 || (_templateObject4 = (0, _taggedTemplateLiteral2["default"])(["\n  cursor: pointer;\n  flex-wrap: wrap;\n  height: auto;\n  justify-content: start;\n  margin-bottom: 2px;\n  padding: 0px 7px 0px 4px;\n  white-space: normal;\n\n  .chickleted-input__placeholder {\n    line-height: 24px;\n    margin: 4px;\n  }\n"])));
var chickletedInput = (0, _styledComponents.css)(_templateObject5 || (_templateObject5 = (0, _taggedTemplateLiteral2["default"])(["\n  ", " ", ";\n"])), function (props) {
  return props.theme.input;
}, function (props) {
  return props.theme.chickletedInputContainer;
});
var chickletedInputLT = (0, _styledComponents.css)(_templateObject6 || (_templateObject6 = (0, _taggedTemplateLiteral2["default"])(["\n  ", " ", ";\n"])), function (props) {
  return props.theme.inputLT;
}, function (props) {
  return props.theme.chickletedInputContainer;
});
var secondaryChickletedInput = (0, _styledComponents.css)(_templateObject7 || (_templateObject7 = (0, _taggedTemplateLiteral2["default"])(["\n  ", " ", ";\n"])), function (props) {
  return props.theme.secondaryInput;
}, function (props) {
  return props.theme.chickletedInputContainer;
});
var inlineInput = (0, _styledComponents.css)(_templateObject8 || (_templateObject8 = (0, _taggedTemplateLiteral2["default"])(["\n  ", " color: ", ";\n  font-size: 13px;\n  letter-spacing: 0.43px;\n  line-height: 18px;\n  height: 24px;\n  font-weight: 400;\n  padding-left: 4px;\n  margin-left: -4px;\n  background-color: transparent;\n  border: 1px solid transparent;\n\n  &:hover {\n    height: 24px;\n    cursor: text;\n    background-color: transparent;\n    border: 1px solid ", ";\n  }\n\n  &:active,\n  .active,\n  &:focus {\n    background-color: transparent;\n    border: 1px solid ", ";\n  }\n"])), function (props) {
  return props.theme.input;
}, function (props) {
  return props.theme.textColor;
}, function (props) {
  return props.theme.labelColor;
}, function (props) {
  return props.theme.inputBorderActiveColor;
});
var switchTrack = (0, _styledComponents.css)(_templateObject9 || (_templateObject9 = (0, _taggedTemplateLiteral2["default"])(["\n  background: ", ";\n  position: absolute;\n  top: 0;\n  left: ", "px;\n  content: '';\n  display: block;\n  width: ", "px;\n  height: ", "px;\n  border-radius: ", ";\n"])), function (props) {
  return props.checked ? props.theme.switchTrackBgdActive : props.theme.switchTrackBgd;
}, function (props) {
  return -props.theme.switchLabelMargin;
}, function (props) {
  return props.theme.switchWidth;
}, function (props) {
  return props.theme.switchHeight;
}, function (props) {
  return props.theme.switchTrackBorderRadius;
});
var switchButton = (0, _styledComponents.css)(_templateObject10 || (_templateObject10 = (0, _taggedTemplateLiteral2["default"])(["\n  transition: ", ";\n  position: absolute;\n  top: ", "px;\n  left: ", "px;\n  content: '';\n  display: block;\n  height: ", "px;\n  width: ", "px;\n  background: ", ";\n  box-shadow: ", ";\n  border-radius: ", ";\n"])), function (props) {
  return props.theme.transition;
}, function (props) {
  return (props.theme.switchHeight - props.theme.switchBtnHeight) / 2;
}, function (props) {
  return (props.checked ? props.theme.switchWidth / 2 : (props.theme.switchHeight - props.theme.switchBtnHeight) / 2) - props.theme.switchLabelMargin;
}, function (props) {
  return props.theme.switchBtnHeight;
}, function (props) {
  return props.theme.switchBtnWidth;
}, function (props) {
  return props.checked ? props.theme.switchBtnBgdActive : props.theme.switchBtnBgd;
}, function (props) {
  return props.theme.switchBtnBoxShadow;
}, function (props) {
  return props.theme.switchBtnBorderRadius;
});
var inputSwitch = (0, _styledComponents.css)(_templateObject11 || (_templateObject11 = (0, _taggedTemplateLiteral2["default"])(["\n  user-select: none;\n  cursor: pointer;\n  line-height: ", "px;\n  font-weight: 500;\n  font-size: 12px;\n  color: ", ";\n  position: relative;\n  display: inline-block;\n  padding-top: 0;\n  padding-right: 0;\n  padding-bottom: 0;\n  padding-left: ", "px;\n\n  &:before {\n    ", ";\n  }\n\n  &:after {\n    ", ";\n  }\n"])), function (props) {
  return props.theme.switchHeight;
}, function (props) {
  return props.theme.labelColor;
}, function (props) {
  return props.theme.switchWidth;
}, function (props) {
  return props.theme.switchTrack;
}, function (props) {
  return props.theme.switchButton;
});

// This is a light version checkbox
var checkboxBox = (0, _styledComponents.css)(_templateObject12 || (_templateObject12 = (0, _taggedTemplateLiteral2["default"])(["\n  display: block;\n  position: absolute;\n  top: 0;\n  left: 0;\n  width: ", "px;\n  height: ", "px;\n  background: ", ";\n  border: 1px solid\n    ", ";\n  border-radius: 2px;\n  content: '';\n"])), function (props) {
  return props.theme.checkboxWidth;
}, function (props) {
  return props.theme.checkboxHeight;
}, function (props) {
  return props.checked ? props.theme.checkboxBoxBgdChecked : props.theme.checkboxBoxBgd;
}, function (props) {
  return props.checked ? props.theme.checkboxBoxBgdChecked : props.theme.checkboxBorderColor;
});
var checkboxCheck = (0, _styledComponents.css)(_templateObject13 || (_templateObject13 = (0, _taggedTemplateLiteral2["default"])(["\n  width: 10px;\n  height: 5px;\n  border-bottom: 2px solid white;\n  border-left: 2px solid white;\n  top: 4px;\n  left: 3px;\n  transform: rotate(-45deg);\n  display: block;\n  position: absolute;\n  opacity: ", ";\n  content: '';\n"])), function (props) {
  return props.checked ? 1 : 0;
});
var inputCheckbox = (0, _styledComponents.css)(_templateObject14 || (_templateObject14 = (0, _taggedTemplateLiteral2["default"])(["\n  display: inline-block;\n  position: relative;\n  padding-left: 32px;\n  margin-bottom: 24px;\n  line-height: 20px;\n  vertical-align: middle;\n  cursor: pointer;\n  font-size: 12px;\n  color: ", ";\n  margin-left: -", "px;\n\n  &:before {\n    ", ";\n  }\n\n  &:after {\n    ", ";\n  }\n"])), function (props) {
  return props.theme.labelColor;
}, function (props) {
  return props.theme.switchLabelMargin;
}, function (props) {
  return props.theme.checkboxBox;
}, function (props) {
  return props.theme.checkboxCheck;
});
var radioTrack = (0, _styledComponents.css)(_templateObject15 || (_templateObject15 = (0, _taggedTemplateLiteral2["default"])(["\n  ", "\n  width: ", "px;\n  height: ", "px;\n  border-radius: ", "px;\n  background-color: ", ";\n  border-color: ", ";\n"])), function (props) {
  return props.theme.checkboxBox;
}, function (props) {
  return props.theme.radioRadius * 2;
}, function (props) {
  return props.theme.radioRadius * 2;
}, function (props) {
  return props.theme.radioBorderRadius;
}, function (props) {
  return props.theme.switchTrackBgd;
}, function (props) {
  return props.theme.radioBorderColor;
});
var radioButton = (0, _styledComponents.css)(_templateObject16 || (_templateObject16 = (0, _taggedTemplateLiteral2["default"])(["\n  border: 0;\n  display: table;\n  top: ", "px;\n  left: ", "px;\n  width: ", "px;\n  height: ", "px;\n  border-radius: ", "px;\n  background-color: ", ";\n"])), function (props) {
  return props.theme.radioRadius - props.theme.radioButtonRadius;
}, function (props) {
  return props.theme.radioRadius - props.theme.radioButtonRadius;
}, function (props) {
  return props.theme.radioButtonRadius * 2;
}, function (props) {
  return props.theme.radioButtonRadius * 2;
}, function (props) {
  return props.theme.radioButtonRadius * 2;
}, function (props) {
  return props.theme.radioButtonBgdColor;
});
var inputRadio = (0, _styledComponents.css)(_templateObject17 || (_templateObject17 = (0, _taggedTemplateLiteral2["default"])(["\n  ", "\n  padding-left: ", "px;\n  margin-bottom: 0;\n  margin-left: 0;\n  line-height: ", "px;\n  color: ", ";\n  cursor: pointer;\n\n  &:before {\n    ", "\n  }\n\n  &:after {\n    ", "\n  }\n"])), function (props) {
  return props.theme.inputCheckbox;
}, function (props) {
  return props.theme.radioRadius * 2 + 8;
}, function (props) {
  return props.theme.radioRadius * 2;
}, function (props) {
  return props.theme.textColorHl;
}, function (props) {
  return props.theme.radioTrack;
}, function (props) {
  return props.theme.radioButton;
});
var secondaryRadio = (0, _styledComponents.css)(_templateObject18 || (_templateObject18 = (0, _taggedTemplateLiteral2["default"])(["\n  ", "\n\n  &:before {\n    ", " background: ", ";\n  }\n\n  &:after {\n    ", "\n    background: ", ";\n  }\n"])), function (props) {
  return props.theme.inputRadio;
}, function (props) {
  return props.theme.radioTrack;
}, function (props) {
  return props.theme.secondarySwitchTrackBgd;
}, function (props) {
  return props.theme.radioButton;
}, function (props) {
  return props.checked ? props.theme.switchBtnBgdActive : props.theme.secondarySwitchBtnBgd;
});
var secondarySwitch = (0, _styledComponents.css)(_templateObject19 || (_templateObject19 = (0, _taggedTemplateLiteral2["default"])(["\n  ", "\n\n  &:before {\n    ", " background: ", ";\n  }\n\n  &:after {\n    ", "\n    background: ", ";\n  }\n"])), function (props) {
  return props.theme.inputSwitch;
}, function (props) {
  return props.theme.switchTrack;
}, function (props) {
  return props.checked ? props.theme.switchTrackBgdActive : props.theme.secondarySwitchTrackBgd;
}, function (props) {
  return props.theme.switchButton;
}, function (props) {
  return props.checked ? props.theme.switchBtnBgdActive : props.theme.secondarySwitchBtnBgd;
});
var dropdownScrollBar = (0, _styledComponents.css)(_templateObject20 || (_templateObject20 = (0, _taggedTemplateLiteral2["default"])(["\n  &::-webkit-scrollbar {\n    height: 10px;\n    width: 10px;\n  }\n\n  &::-webkit-scrollbar-corner {\n    background: ", ";\n  }\n\n  &::-webkit-scrollbar-track {\n    background: ", ";\n  }\n\n  &::-webkit-scrollbar-thumb {\n    border-radius: 10px;\n    background: ", ";\n    border: 3px solid ", ";\n  }\n\n  &:vertical:hover {\n    background: ", ";\n    cursor: pointer;\n  }\n"])), function (props) {
  return props.theme.dropdownListBgd;
}, function (props) {
  return props.theme.dropdownListBgd;
}, function (props) {
  return props.theme.labelColor;
}, function (props) {
  return props.theme.dropdownListBgd;
}, function (props) {
  return props.theme.textColorHl;
});
var dropdownScrollBarLT = (0, _styledComponents.css)(_templateObject21 || (_templateObject21 = (0, _taggedTemplateLiteral2["default"])(["\n  ", " ::-webkit-scrollbar-corner {\n    background: ", ";\n  }\n\n  &::-webkit-scrollbar-track {\n    background: ", ";\n  }\n\n  &::-webkit-scrollbar-thumb {\n    border-radius: 10px;\n    background: ", ";\n    border: 3px solid ", ";\n  }\n\n  &:vertical:hover {\n    background: ", ";\n    cursor: pointer;\n  }\n"])), dropdownScrollBar, function (props) {
  return props.theme.dropdownListBgdLT;
}, function (props) {
  return props.theme.dropdownListBgdLT;
}, function (props) {
  return props.theme.scrollbarThumbColorLT;
}, function (props) {
  return props.theme.dropdownListBgdLT;
}, function (props) {
  return props.theme.scrollbarThumbColorHoverLT;
});
var dropdownListAnchor = (0, _styledComponents.css)(_templateObject22 || (_templateObject22 = (0, _taggedTemplateLiteral2["default"])(["\n  color: ", ";\n  padding-left: 3px;\n  font-size: ", ";\n  line-height: ", "px;\n  overflow: hidden;\n  text-overflow: ellipsis;\n\n  &.disabled {\n    color: ", ";\n  }\n"])), function (props) {
  return props.theme.selectColor;
}, function (props) {
  return props.theme.selectFontSize;
}, function (props) {
  return props.theme.dropdownListLineHeight;
}, function (props) {
  return props.theme.subtextColor;
});
var dropdownListAnchorLT = (0, _styledComponents.css)(_templateObject23 || (_templateObject23 = (0, _taggedTemplateLiteral2["default"])(["\n  ", "\n  color: ", ";\n"])), dropdownListAnchor, function (props) {
  return props.theme.selectColorLT;
});
var dropdownListSize = (0, _styledComponents.css)(_templateObject24 || (_templateObject24 = (0, _taggedTemplateLiteral2["default"])(["\n  font-size: 11px;\n  padding: 3px 9px;\n  font-weight: 500;\n  white-space: nowrap;\n"])));
var dropdownListItem = (0, _styledComponents.css)(_templateObject25 || (_templateObject25 = (0, _taggedTemplateLiteral2["default"])(["\n  ", " &.hover,\n  &:hover {\n    cursor: pointer;\n    background-color: ", ";\n\n    .list__item__anchor {\n      color: ", ";\n    }\n  }\n  &.selected {\n    background-color: ", ";\n  }\n"])), dropdownListSize, function (props) {
  return props.theme.dropdownListHighlightBg;
}, function (props) {
  return props.theme.textColorHl;
}, function (props) {
  return props.theme.dropdownListHighlightBg;
});
var dropdownListItemLT = (0, _styledComponents.css)(_templateObject26 || (_templateObject26 = (0, _taggedTemplateLiteral2["default"])(["\n  ", "\n  color: ", ";\n\n  &.hover,\n  &:hover {\n    cursor: pointer;\n    color: ", ";\n    background-color: ", ";\n\n    .list__item__anchor {\n      color: ", ";\n    }\n  }\n  &.selected {\n    background-color: ", ";\n  }\n"])), dropdownListSize, function (props) {
  return props.theme.textColorLT;
}, function (props) {
  return props.theme.textColorHlLT;
}, function (props) {
  return props.theme.dropdownListHighlightBgLT;
}, function (props) {
  return props.theme.selectColorLT;
}, function (props) {
  return props.theme.dropdownListHighlightBgLT;
});
var dropdownListHeader = (0, _styledComponents.css)(_templateObject27 || (_templateObject27 = (0, _taggedTemplateLiteral2["default"])(["\n  font-size: 11px;\n  padding: 5px 9px;\n  color: ", ";\n"])), function (props) {
  return props.theme.labelColor;
});
var dropdownListSection = (0, _styledComponents.css)(_templateObject28 || (_templateObject28 = (0, _taggedTemplateLiteral2["default"])(["\n  padding: 0 0 4px 0;\n  margin-bottom: 4px;\n  border-bottom: 1px solid ", ";\n"])), function (props) {
  return props.theme.labelColor;
});
var dropdownList = (0, _styledComponents.css)(_templateObject29 || (_templateObject29 = (0, _taggedTemplateLiteral2["default"])(["\n  overflow-y: auto;\n  max-height: 280px;\n  box-shadow: ", ";\n  border-radius: 2px;\n\n  .list__section {\n    ", ";\n  }\n  .list__header {\n    ", ";\n  }\n\n  .list__item {\n    ", ";\n  }\n\n  .list__item__anchor {\n    ", ";\n  }\n\n  ", ";\n"])), function (props) {
  return props.theme.dropdownListShadow;
}, function (props) {
  return props.theme.dropdownListSection;
}, function (props) {
  return props.theme.dropdownListHeader;
}, function (props) {
  return props.theme.dropdownListItem;
}, function (props) {
  return props.theme.dropdownListAnchor;
}, function (props) {
  return props.theme.dropdownScrollBar;
});
var dropdownListLT = (0, _styledComponents.css)(_templateObject30 || (_templateObject30 = (0, _taggedTemplateLiteral2["default"])(["\n  ", " .list__item {\n    ", ";\n  }\n\n  .list__item__anchor {\n    ", ";\n  }\n\n  ", ";\n"])), dropdownList, function (props) {
  return props.theme.dropdownListItemLT;
}, function (props) {
  return props.theme.dropdownListAnchorLT;
}, function (props) {
  return props.theme.dropdownScrollBarLT;
});
var sidePanelScrollBar = (0, _styledComponents.css)(_templateObject31 || (_templateObject31 = (0, _taggedTemplateLiteral2["default"])(["\n  &::-webkit-scrollbar {\n    height: ", "px;\n    width: ", "px;\n  }\n\n  &::-webkit-scrollbar-corner {\n    background: ", ";\n  }\n\n  &::-webkit-scrollbar-track {\n    background: ", ";\n  }\n\n  &::-webkit-scrollbar-thumb {\n    border-radius: 10px;\n    background: ", ";\n    border: 3px solid ", ";\n\n    &:hover {\n      background: ", ";\n      cursor: pointer;\n    }\n  }\n"])), function (props) {
  return props.theme.sidePanelScrollBarHeight;
}, function (props) {
  return props.theme.sidePanelScrollBarWidth;
}, function (props) {
  return props.theme.sidePanelBg;
}, function (props) {
  return props.theme.sidePanelBg;
}, function (props) {
  return props.theme.panelBackgroundHover;
}, function (props) {
  return props.theme.sidePanelBg;
}, function (props) {
  return props.theme.labelColor;
});
var panelDropdownScrollBar = (0, _styledComponents.css)(_templateObject32 || (_templateObject32 = (0, _taggedTemplateLiteral2["default"])(["\n  &::-webkit-scrollbar {\n    height: 10px;\n    width: 10px;\n  }\n\n  &::-webkit-scrollbar-corner {\n    background: ", ";\n  }\n\n  &::-webkit-scrollbar-track {\n    background: ", ";\n  }\n\n  &::-webkit-scrollbar-thumb {\n    border-radius: 10px;\n    background: ", ";\n    border: 3px solid ", ";\n    &:hover {\n      background: ", ";\n      cursor: pointer;\n    }\n  }\n"])), function (props) {
  return props.theme.panelBackground;
}, function (props) {
  return props.theme.panelBackground;
}, function (props) {
  return props.theme.panelBackgroundHover;
}, function (props) {
  return props.theme.panelBackground;
}, function (props) {
  return props.theme.labelColor;
});
var scrollBar = (0, _styledComponents.css)(_templateObject33 || (_templateObject33 = (0, _taggedTemplateLiteral2["default"])(["\n  &::-webkit-scrollbar {\n    height: 10px;\n    width: 10px;\n  }\n\n  &::-webkit-scrollbar-corner {\n    background: ", ";\n  }\n\n  &::-webkit-scrollbar-track {\n    background: ", ";\n  }\n\n  &::-webkit-scrollbar-thumb {\n    border-radius: 10px;\n    background: ", ";\n    border: 3px solid ", ";\n\n    :vertical:hover {\n      background: ", ";\n      cursor: pointer;\n    }\n\n    :horizontal:hover {\n      background: ", ";\n      cursor: pointer;\n    }\n  }\n"])), function (props) {
  return props.theme.panelBackground;
}, function (props) {
  return props.theme.panelBackground;
}, function (props) {
  return props.theme.labelColor;
}, function (props) {
  return props.theme.panelBackground;
}, function (props) {
  return props.theme.textColorHl;
}, function (props) {
  return props.theme.textColorHl;
});
var modalScrollBar = exports.modalScrollBar = (0, _styledComponents.css)(_templateObject34 || (_templateObject34 = (0, _taggedTemplateLiteral2["default"])(["\n  &::-webkit-scrollbar {\n    width: 14px;\n    height: 16px;\n  }\n\n  &::-webkit-scrollbar-track {\n    background: white;\n  }\n  &::-webkit-scrollbar-track:horizontal {\n    background: ", ";\n  }\n  &::-webkit-scrollbar-thumb {\n    background: ", ";\n    border: 4px solid white;\n  }\n\n  &::-webkit-scrollbar-corner {\n    background: ", ";\n  }\n\n  &::-webkit-scrollbar-thumb:hover {\n    background: ", ";\n  }\n\n  &::-webkit-scrollbar-thumb:vertical {\n    border-radius: 7px;\n  }\n\n  &::-webkit-scrollbar-thumb:horizontal {\n    border-radius: 9px;\n    border: 4px solid ", ";\n  }\n"])), function (props) {
  return props.theme.textColorHl;
}, function (props) {
  return props.theme.scrollbarThumbColorLT;
}, function (props) {
  return props.theme.textColorHl;
}, function (props) {
  return props.theme.scrollbarThumbColorHoverLT;
}, function (props) {
  return props.theme.textColorHl;
});
var theme = exports.theme = _objectSpread(_objectSpread({}, _constants.DIMENSIONS), {}, {
  // templates
  input: input,
  inputLT: inputLT,
  inlineInput: inlineInput,
  chickletedInput: chickletedInput,
  chickletedInputLT: chickletedInputLT,
  chickletedInputContainer: chickletedInputContainer,
  secondaryChickletedInput: secondaryChickletedInput,
  borderColor: borderColor,
  borderColorLT: borderColorLT,
  secondaryInput: secondaryInput,
  dropdownScrollBar: dropdownScrollBar,
  dropdownScrollBarLT: dropdownScrollBarLT,
  dropdownList: dropdownList,
  dropdownListLT: dropdownListLT,
  dropdownListItem: dropdownListItem,
  dropdownListItemLT: dropdownListItemLT,
  dropdownListAnchor: dropdownListAnchor,
  dropdownListAnchorLT: dropdownListAnchorLT,
  dropdownListHeader: dropdownListHeader,
  dropdownListSection: dropdownListSection,
  dropdownListShadow: dropdownListShadow,
  dropdownWrapperZ: dropdownWrapperZ,
  dropdownWapperMargin: dropdownWapperMargin,
  dndOverBackgroundColor: dndOverBackgroundColor,
  modalScrollBar: modalScrollBar,
  scrollBar: scrollBar,
  sidePanelScrollBar: sidePanelScrollBar,
  inputSwitch: inputSwitch,
  secondarySwitch: secondarySwitch,
  switchTrack: switchTrack,
  switchButton: switchButton,
  inputCheckbox: inputCheckbox,
  inputRadio: inputRadio,
  checkboxBox: checkboxBox,
  checkboxCheck: checkboxCheck,
  // Transitions
  transition: transition,
  transitionFast: transitionFast,
  transitionSlow: transitionSlow,
  // styles
  activeColor: activeColor,
  activeColorHover: activeColorHover,
  borderRadius: borderRadius,
  boxShadow: boxShadow,
  errorColor: errorColor,
  dropdownListHighlightBg: dropdownListHighlightBg,
  dropdownListHighlightBgLT: dropdownListHighlightBgLT,
  dropdownListBgd: dropdownListBgd,
  toolbarItemBgdHover: toolbarItemBgdHover,
  toolbarItemBorderHover: toolbarItemBorderHover,
  toolbarItemIconHover: toolbarItemIconHover,
  toolbarItemBorderRaddius: toolbarItemBorderRaddius,
  dropdownListBgdLT: dropdownListBgdLT,
  dropdownListBorderTop: dropdownListBorderTop,
  dropdownListBorderTopLT: dropdownListBorderTopLT,
  dropdownListLineHeight: dropdownListLineHeight,
  labelColor: labelColor,
  labelColorLT: labelColorLT,
  labelHoverColor: labelHoverColor,
  mapPanelBackgroundColor: mapPanelBackgroundColor,
  mapPanelHeaderBackgroundColor: mapPanelHeaderBackgroundColor,
  // Select
  selectActiveBorderColor: selectActiveBorderColor,
  selectBackground: selectBackground,
  selectBackgroundLT: selectBackgroundLT,
  selectBackgroundHover: selectBackgroundHover,
  selectBackgroundHoverLT: selectBackgroundHoverLT,
  selectBorder: selectBorder,
  selectBorderColor: selectBorderColor,
  selectBorderRadius: selectBorderRadius,
  selectBorderColorLT: selectBorderColorLT,
  selectColor: selectColor,
  selectColorPlaceHolder: selectColorPlaceHolder,
  selectColorPlaceHolderLT: selectColorPlaceHolderLT,
  selectFontSize: selectFontSize,
  selectFontWeight: selectFontWeight,
  selectColorLT: selectColorLT,
  selectFontWeightBold: selectFontWeightBold,
  panelTabColor: panelTabColor,
  // Input
  inputBgd: inputBgd,
  inputBgdHover: inputBgdHover,
  inputBgdActive: inputBgdActive,
  inputBgdActiveLT: inputBgdActiveLT,
  inputBoxHeight: inputBoxHeight,
  inputBoxHeightSmall: inputBoxHeightSmall,
  inputBoxHeightTiny: inputBoxHeightTiny,
  inputBorderColor: inputBorderColor,
  inputBorderActiveColor: inputBorderActiveColor,
  inputBorderHoverColor: inputBorderHoverColor,
  inputBorderRadius: inputBorderRadius,
  inputColor: inputColor,
  inputPadding: inputPadding,
  inputPaddingSmall: inputPaddingSmall,
  inputPaddingTiny: inputPaddingTiny,
  inputFontSize: inputFontSize,
  inputFontSizeSmall: inputFontSizeSmall,
  inputFontWeight: inputFontWeight,
  inputPlaceholderColor: inputPlaceholderColor,
  inputPlaceholderColorLT: inputPlaceholderColorLT,
  inputPlaceholderFontWeight: inputPlaceholderFontWeight,
  inputBoxShadow: inputBoxShadow,
  inputBoxShadowActive: inputBoxShadowActive,
  inputBoxShadowActiveLT: inputBoxShadowActiveLT,
  secondaryInputBgd: secondaryInputBgd,
  secondaryInputBgdHover: secondaryInputBgdHover,
  secondaryInputBgdActive: secondaryInputBgdActive,
  secondaryInputColor: secondaryInputColor,
  secondaryInputBorderColor: secondaryInputBorderColor,
  secondaryInputBorderActiveColor: secondaryInputBorderActiveColor,
  dropdownSelectHeight: dropdownSelectHeight,
  // Switch
  switchWidth: switchWidth,
  switchHeight: switchHeight,
  switchTrackBgd: switchTrackBgd,
  switchTrackBgdActive: switchTrackBgdActive,
  switchTrackBorderRadius: switchTrackBorderRadius,
  switchBtnBgd: switchBtnBgd,
  switchBtnBgdActive: switchBtnBgdActive,
  switchBtnBoxShadow: switchBtnBoxShadow,
  switchBtnBorderRadius: switchBtnBorderRadius,
  switchBtnWidth: switchBtnWidth,
  switchBtnHeight: switchBtnHeight,
  switchLabelMargin: switchLabelMargin,
  secondarySwitchTrackBgd: secondarySwitchTrackBgd,
  secondarySwitchBtnBgd: secondarySwitchBtnBgd,
  // Checkbox
  checkboxWidth: checkboxWidth,
  checkboxHeight: checkboxHeight,
  checkboxMargin: checkboxMargin,
  checkboxBorderColor: checkboxBorderColor,
  checkboxBorderRadius: checkboxBorderRadius,
  checkboxBorderColorLT: checkboxBorderColorLT,
  checkboxBoxBgd: checkboxBoxBgd,
  checkboxBoxBgdChecked: checkboxBoxBgdChecked,
  // Radio
  radioRadius: radioRadius,
  radioBorderRadius: radioBorderRadius,
  radioBorderColor: radioBorderColor,
  radioButtonRadius: radioButtonRadius,
  radioButtonBgdColor: radioButtonBgdColor,
  radioTrack: radioTrack,
  radioButton: radioButton,
  secondaryRadio: secondaryRadio,
  // Button
  btnFontFamily: btnFontFamily,
  primaryBtnBgd: primaryBtnBgd,
  primaryBtnActBgd: primaryBtnActBgd,
  primaryBtnColor: primaryBtnColor,
  primaryBtnActColor: primaryBtnActColor,
  primaryBtnBgdHover: primaryBtnBgdHover,
  primaryBtnRadius: primaryBtnRadius,
  primaryBtnFontSizeDefault: primaryBtnFontSizeDefault,
  primaryBtnFontSizeSmall: primaryBtnFontSizeSmall,
  primaryBtnFontSizeLarge: primaryBtnFontSizeLarge,
  primaryBtnBorder: primaryBtnBorder,
  secondaryBtnBgd: secondaryBtnBgd,
  secondaryBtnActBgd: secondaryBtnActBgd,
  secondaryBtnBgdHover: secondaryBtnBgdHover,
  secondaryBtnColor: secondaryBtnColor,
  secondaryBtnActColor: secondaryBtnActColor,
  secondaryBtnBorder: secondaryBtnBorder,
  negativeBtnBgd: negativeBtnBgd,
  negativeBtnActBgd: negativeBtnActBgd,
  negativeBtnBgdHover: negativeBtnBgdHover,
  negativeBtnBorder: negativeBtnBorder,
  negativeBtnColor: negativeBtnColor,
  negativeBtnActColor: negativeBtnActColor,
  linkBtnBgd: linkBtnBgd,
  linkBtnActBgd: linkBtnActBgd,
  linkBtnColor: linkBtnColor,
  linkBtnActColor: linkBtnActColor,
  linkBtnActBgdHover: linkBtnActBgdHover,
  linkBtnBorder: linkBtnBorder,
  floatingBtnBgd: floatingBtnBgd,
  floatingBtnActBgd: floatingBtnActBgd,
  floatingBtnBgdHover: floatingBtnBgdHover,
  floatingBtnBorder: floatingBtnBorder,
  floatingBtnBorderHover: floatingBtnBorderHover,
  floatingBtnColor: floatingBtnColor,
  floatingBtnActColor: floatingBtnActColor,
  floatingBtnRadius: floatingBtnRadius,
  ctaBtnBgd: ctaBtnBgd,
  ctaBtnBgdHover: ctaBtnBgdHover,
  ctaBtnActBgd: ctaBtnActBgd,
  ctaBtnColor: ctaBtnColor,
  ctaBtnActColor: ctaBtnActColor,
  selectionBtnBgd: selectionBtnBgd,
  selectionBtnActBgd: selectionBtnActBgd,
  selectionBtnColor: selectionBtnColor,
  selectionBtnActColor: selectionBtnActColor,
  selectionBtnBgdHover: selectionBtnBgdHover,
  selectionBtnBorder: selectionBtnBorder,
  selectionBtnBorderColor: selectionBtnBorderColor,
  selectionBtnBorderActColor: selectionBtnBorderActColor,
  scrollbarThumbColorLT: scrollbarThumbColorLT,
  scrollbarThumbColorHoverLT: scrollbarThumbColorHoverLT,
  // Modal
  modalTitleColor: modalTitleColor,
  modalTitleFontSize: modalTitleFontSize,
  modalTitleFontSizeSmaller: modalTitleFontSizeSmaller,
  modalFooterBgd: modalFooterBgd,
  modalImagePlaceHolder: modalImagePlaceHolder,
  modalPadding: modalPadding,
  modalDialogBgd: modalDialogBgd,
  modalDialogColor: modalDialogColor,
  modalLateralPadding: modalLateralPadding,
  modalPortableLateralPadding: modalPortableLateralPadding,
  modalOverLayZ: modalOverLayZ,
  modalOverlayBgd: modalOverlayBgd,
  modalContentZ: modalContentZ,
  modalFooterZ: modalFooterZ,
  modalTitleZ: modalTitleZ,
  modalButtonZ: modalButtonZ,
  modalDropdownBackground: modalDropdownBackground,
  // Side Panel
  sidePanelBg: sidePanelBg,
  sidePanelInnerPadding: sidePanelInnerPadding,
  sideBarCloseBtnBgd: sideBarCloseBtnBgd,
  sideBarCloseBtnColor: sideBarCloseBtnColor,
  sideBarCloseBtnBgdHover: sideBarCloseBtnBgdHover,
  sidePanelHeaderBg: sidePanelHeaderBg,
  sidePanelHeaderBorder: sidePanelHeaderBorder,
  sidePanelScrollBarWidth: sidePanelScrollBarWidth,
  sidePanelScrollBarHeight: sidePanelScrollBarHeight,
  sidePanelTitleFontsize: sidePanelTitleFontsize,
  sidePanelTitleLineHeight: sidePanelTitleLineHeight,
  panelHeaderBorderRadius: panelHeaderBorderRadius,
  sidePanelBorder: sidePanelBorder,
  sidePanelBorderColor: sidePanelBorderColor,
  layerConfigGroupLabelLabelFontSize: layerConfigGroupLabelLabelFontSize,
  layerConfigGroupMarginBottom: layerConfigGroupMarginBottom,
  layerConfigGroupPaddingLeft: layerConfigGroupPaddingLeft,
  // Side Panel Panel
  chickletBgd: chickletBgd,
  chickletBgdLT: chickletBgdLT,
  panelBackground: panelBackground,
  panelContentBackground: panelContentBackground,
  panelBackgroundHover: panelBackgroundHover,
  panelBackgroundLT: panelBackgroundLT,
  panelToggleMarginRight: panelToggleMarginRight,
  panelToggleBottomPadding: panelToggleBottomPadding,
  panelBoxShadow: panelBoxShadow,
  panelBorderRadius: panelBorderRadius,
  panelBorder: panelBorder,
  panelBorderColor: panelBorderColor,
  panelBorderLT: panelBorderLT,
  panelHeaderIcon: panelHeaderIcon,
  panelHeaderIconActive: panelHeaderIconActive,
  panelHeaderIconHover: panelHeaderIconHover,
  panelHeaderHeight: panelHeaderHeight,
  layerPanelHeaderHeight: layerPanelHeaderHeight,
  panelDropdownScrollBar: panelDropdownScrollBar,
  layerTypeIconSizeL: layerTypeIconSizeL,
  layerTypeIconPdL: layerTypeIconPdL,
  layerTypeIconSizeSM: layerTypeIconSizeSM,
  layerPanelToggleOptionColor: layerPanelToggleOptionColor,
  layerPanelToggleOptionColorActive: layerPanelToggleOptionColorActive,
  // Text
  fontFamily: fontFamily,
  fontWeight: fontWeight,
  fontSize: fontSize,
  lineHeight: lineHeight,
  textColor: textColor,
  textColorLT: textColorLT,
  dataTableTextColor: dataTableTextColor,
  textColorHl: textColorHl,
  titleTextColor: titleTextColor,
  subtextColor: subtextColor,
  subtextColorLT: subtextColorLT,
  subtextColorActive: subtextColorActive,
  fontWhiteColor: fontWhiteColor,
  panelToggleBorderColor: panelToggleBorderColor,
  panelTabWidth: panelTabWidth,
  textTruncate: textTruncate,
  titleColorLT: titleColorLT,
  tooltipBg: tooltipBg,
  tooltipBgLT: tooltipBgLT,
  tooltipColor: tooltipColor,
  tooltipColorLT: tooltipColorLT,
  tooltipBoxShadow: tooltipBoxShadow,
  tooltipFontSize: tooltipFontSize,
  logoColor: logoColor,
  // Sidepanel divider
  sidepanelDividerBorder: sidepanelDividerBorder,
  sidepanelDividerMargin: sidepanelDividerMargin,
  sidepanelDividerHeight: sidepanelDividerHeight,
  // Bottom Panel
  bottomInnerPdSide: bottomInnerPdSide,
  bottomInnerPdVert: bottomInnerPdVert,
  bottomPanelGap: bottomPanelGap,
  bottomPanelGapPalm: bottomPanelGapPalm,
  bottomWidgetPaddingTop: bottomWidgetPaddingTop,
  bottomWidgetPaddingRight: bottomWidgetPaddingRight,
  bottomWidgetPaddingBottom: bottomWidgetPaddingBottom,
  bottomWidgetPaddingLeft: bottomWidgetPaddingLeft,
  bottomWidgetBgd: bottomWidgetBgd,
  // Slider
  sliderBarColor: sliderBarColor,
  sliderBarBgd: sliderBarBgd,
  sliderBarHoverColor: sliderBarHoverColor,
  sliderBarRadius: sliderBarRadius,
  sliderBarHeight: sliderBarHeight,
  sliderHandleHeight: sliderHandleHeight,
  sliderHandleWidth: sliderHandleWidth,
  sliderHandleColor: sliderHandleColor,
  sliderHandleTextColor: sliderHandleTextColor,
  sliderInactiveBorderColor: sliderInactiveBorderColor,
  sliderBorderRadius: sliderBorderRadius,
  sliderHandleHoverColor: sliderHandleHoverColor,
  sliderHandleAfterContent: sliderHandleAfterContent,
  sliderHandleShadow: sliderHandleShadow,
  sliderInputHeight: sliderInputHeight,
  sliderInputWidth: sliderInputWidth,
  sliderMarginTopIsTime: sliderMarginTopIsTime,
  sliderMarginTop: sliderMarginTop,
  sliderMarginBottom: sliderMarginBottom,
  // Custom Range Input
  customRangeInputWidth: customRangeInputWidth,
  // Geocoder
  geocoderWidth: geocoderWidth,
  geocoderTop: geocoderTop,
  geocoderRight: geocoderRight,
  geocoderInputHeight: geocoderInputHeight,
  // Map Control
  mapControlTop: mapControlTop,
  // Plot
  rangeBrushBgd: rangeBrushBgd,
  histogramFillInRange: histogramFillInRange,
  histogramFillOutRange: histogramFillOutRange,
  histogramOverlayColor: histogramOverlayColor,
  histogramBreakLineColor: histogramBreakLineColor,
  axisFontSize: axisFontSize,
  axisFontColor: axisFontColor,
  timeTitleFontSize: timeTitleFontSize,
  rangePlotMargin: rangePlotMargin,
  rangePlotMarginLarge: rangePlotMarginLarge,
  rangePlotH: rangePlotH,
  rangePlotHLarge: rangePlotHLarge,
  rangePlotHLargePalm: rangePlotHLargePalm,
  rangePlotContainerH: rangePlotContainerH,
  rangePlotContainerHLarge: rangePlotContainerHLarge,
  rangePlotContainerHLargePalm: rangePlotContainerHLargePalm,
  // Notifications
  notificationColors: notificationColors,
  notificationPanelWidth: notificationPanelWidth,
  notificationPanelItemWidth: notificationPanelItemWidth,
  notificationPanelItemHeight: notificationPanelItemHeight,
  // Data Table
  headerRowHeight: headerRowHeight,
  headerRowWStatsHeight: headerRowWStatsHeight,
  headerStatsControlHeight: headerStatsControlHeight,
  rowHeight: rowHeight,
  headerPaddingTop: headerPaddingTop,
  headerPaddingBottom: headerPaddingBottom,
  cellPaddingSide: cellPaddingSide,
  edgeCellPaddingSide: edgeCellPaddingSide,
  cellFontSize: cellFontSize,
  gridPaddingSide: gridPaddingSide,
  optionButtonColor: optionButtonColor,
  headerCellBackground: headerCellBackground,
  headerCellBorderColor: headerCellBorderColor,
  headerCellStatsBackground: headerCellStatsBackground,
  headerCellStatsControlBackground: headerCellStatsControlBackground,
  headerCellIconColor: headerCellIconColor,
  cellBorderColor: cellBorderColor,
  evenRowBackground: evenRowBackground,
  oddRowBackground: oddRowBackground,
  pinnedGridBorderColor: pinnedGridBorderColor,
  // time display
  timeDisplayBorderRadius: timeDisplayBorderRadius,
  timeDisplayHeight: timeDisplayHeight,
  timeDisplayMinWidth: timeDisplayMinWidth,
  timeDisplayOpacity: timeDisplayOpacity,
  timeDisplayPadding: timeDisplayPadding,
  // export map
  exportIntraSectionMargin: exportIntraSectionMargin,
  // Action Panel
  actionPanelWidth: actionPanelWidth,
  actionPanelHeight: actionPanelHeight,
  // Breakpoints
  breakPoints: breakPoints,
  // progressbar
  progressBarColor: progressBarColor,
  progressBarTrackColor: progressBarTrackColor,
  // layerConfigGroupLabel
  layerConfigGroupLabelBorderLeft: layerConfigGroupLabelBorderLeft,
  layerConfigGroupLabelMargin: layerConfigGroupLabelMargin,
  layerConfigGroupLabelPadding: layerConfigGroupLabelPadding,
  layerConfigGroupColor: layerConfigGroupColor,
  // layerConfigGroupLabel label
  layerConfigGroupLabelLabelMargin: layerConfigGroupLabelLabelMargin,
  // StyledConfigGroupHeader
  styledConfigGroupHeaderBorder: styledConfigGroupHeaderBorder,
  // layerConfigurator
  layerConfiguratorBorder: layerConfiguratorBorder,
  layerConfiguratorBorderColor: layerConfiguratorBorderColor,
  layerConfiguratorMargin: layerConfiguratorMargin,
  layerConfiguratorPadding: layerConfiguratorPadding,
  // Styled token
  fieldTokenRightMargin: fieldTokenRightMargin,
  fieldTokenHeight: fieldTokenHeight,
  fieldTokenWidth: fieldTokenWidth,
  // COLORS
  BLUE2: 'rgba(85, 88, 219, 0.2)',
  // AI Assistant Panel
  aiAssistantPanelWidth: aiAssistantPanelWidth,
  // Effect panel
  effectPanelWidth: effectPanelWidth,
  effectPanelHeight: effectPanelHeight,
  effectPanelPaddingSide: effectPanelPaddingSide,
  effectPanelPaddingTop: effectPanelPaddingTop,
  rightPanelMarginTop: rightPanelMarginTop,
  rightPanelMarginRight: rightPanelMarginRight,
  effectPanelAddEffectFontFamily: effectPanelAddEffectFontFamily,
  // effect type selector
  effectTypeIconMarginSide: effectTypeIconMarginSide,
  effectTypeIconSizeL: effectTypeIconSizeL,
  effectTypeIconBgHoverColor: effectTypeIconBgHoverColor,
  // effectConfigurator
  effectConfiguratorMargin: effectConfiguratorMargin,
  effectConfiguratorPadding: effectConfiguratorPadding,
  effectPanelTextMain: effectPanelTextMain,
  effectPanelTextSecondary1: effectPanelTextSecondary1,
  effectPanelTextSecondary2: effectPanelTextSecondary2,
  effectPanelTextSecondary3: effectPanelTextSecondary3,
  effectPanelElementColor: effectPanelElementColor,
  effectPanelElementColorActive: effectPanelElementColorActive,
  effectPanelElementColorHovered: effectPanelElementColorHovered,
  effectPanelElementColorSun: effectPanelElementColorSun,
  effectPanelElementColorSelected: effectPanelElementColorSelected
});
var themeLT = exports.themeLT = _objectSpread(_objectSpread({}, theme), {}, {
  // template
  activeColor: activeColorLT,
  input: inputLT,
  textColor: textColorLT,
  sidePanelBg: '#FFFFFF',
  selectColor: selectColorLT,
  titleTextColor: '#000000',
  sidePanelHeaderBg: '#F7F7F7',
  subtextColorActive: activeColorLT,
  tooltipBg: tooltipBgLT,
  tooltipColor: tooltipColorLT,
  dropdownListBgd: '#FFFFFF',
  toolbarItemBgdHover: '#F7F7F7',
  textColorHl: activeColorLT,
  inputBgd: '#F7F7F7',
  inputBgdHover: '#FFFFFF',
  inputBgdActive: '#FFFFFF',
  inputBgdActiveLT: '#FFFFFF',
  dropdownListHighlightBg: '#F0F0F0',
  toolbarItemIconHover: activeColorLT,
  panelBackground: '#F7F7F7',
  panelContentBackground: '#F7F7F7',
  bottomWidgetBgd: '#F7F7F7',
  panelBackgroundHover: '#F7F7F7',
  panelBorderColor: '#D3D8E0',
  panelHeaderIconActive: '#000000',
  panelHeaderIconHover: '#000000',
  sideBarCloseBtnBgd: '#F7F7F7',
  sideBarCloseBtnColor: textColorLT,
  sideBarCloseBtnBgdHover: '#F7F7F7',
  secondaryInputBgd: '#F7F7F7',
  secondaryInputBgdActive: '#F7F7F7',
  secondaryInputBgdHover: '#FFFFFF',
  secondaryInputBorderActiveColor: '#000000',
  secondaryInputBorderColor: 'none',
  secondaryInputColor: '#545454',
  chickletBgd: '#F7F7F7',
  mapPanelBackgroundColor: '#FFFFFF',
  mapPanelHeaderBackgroundColor: '#F7F7F7',
  sliderBarColor: '#A0A7B4',
  sliderBarBgd: '#D3D8E0',
  sliderHandleColor: '#F7F7F7',
  sliderInactiveBorderColor: '#F7F7F7',
  sliderHandleTextColor: '#F7F7F7',
  sliderHandleHoverColor: '#F7F7F7',
  subtextColor: subtextColorLT,
  switchBtnBgd: '#F7F7F7',
  secondarySwitchBtnBgd: '#F7F7F7',
  secondarySwitchTrackBgd: '#D3D8E0',
  switchBtnBgdActive: '#F7F7F7',
  switchTrackBgd: '#D3D8E0',
  switchTrackBgdActive: activeColorLT,
  // button switch background and hover color
  primaryBtnBgd: primaryBtnActBgd,
  primaryBtnActBgd: primaryBtnBgd,
  primaryBtnBgdHover: primaryBtnBgd,
  secondaryBtnBgd: secondaryBtnActBgd,
  secondaryBtnActBgd: secondaryBtnBgd,
  secondaryBtnBgdHover: secondaryBtnBgd,
  floatingBtnBgd: '#F7F7F7',
  floatingBtnActBgd: '#F7F7F7',
  floatingBtnBgdHover: '#F7F7F7',
  floatingBtnColor: subtextColor,
  floatingBtnActColor: activeColorLT,
  linkBtnActColor: textColorLT,
  rangeBrushBgd: '#D3D8E0',
  histogramFillInRange: activeColorLT,
  histogramFillOutRange: '#A0A7B4',
  axisFontColor: textColorLT
});
var themeBS = exports.themeBS = _objectSpread(_objectSpread({}, theme), {}, {
  activeColor: '#E2E2E2',
  dropdownListBgd: '#FFFFFF',
  toolbarItemBgdHover: '#F7F7F7',
  dropdownListBorderTop: 'none',
  dropdownListHighlightBg: '#F6F6F6',
  toolbarItemIconHover: '#000000',
  inputBgd: '#E2E2E2',
  inputBgdActive: '#E2E2E2',
  inputBgdHover: 'inherit',
  inputBorderActiveColor: '#000000',
  inputColor: '#000000',
  chickletBgd: '#E2E2E2',
  panelBackground: '#FFFFFF',
  panelBackgroundHover: '#EEEEEE',
  panelContentBackground: '#FFFFFF',
  bottomWidgetBgd: '#F7F7F7',
  panelHeaderIconActive: '#000000',
  panelHeaderIconHover: '#000000',
  panelBorderColor: '#E2E2E2',
  primaryBtnBgd: '#E2E2E2',
  primaryBtnBgdHover: '#333333',
  primaryBtnColor: '#000000',
  secondaryBtnActBgd: '#EEEEEE',
  secondaryBtnActColor: '#000000',
  secondaryBtnBgd: '#E2E2E2',
  secondaryBtnBgdHover: '#CBCBCB',
  ctnBtnBgd: '#E2E2E2',
  ctnBtnBgdHover: '333333',
  ctnBtnColor: '#000000',
  ctnBtnActColor: '#000000',
  sideBarCloseBtnBgd: '#E2E2E2',
  sideBarCloseBtnColor: '#000000',
  sideBarCloseBtnBgdHover: '#FFFFFF',
  floatingBtnBgd: '#FFFFFF',
  floatingBtnActBgd: '#EEEEEE',
  floatingBtnBgdHover: '#EEEEEE',
  floatingBtnColor: '#757575',
  floatingBtnActColor: '#000000',
  secondaryBtnColor: '#000000',
  secondaryInputBgd: '#F6F6F6',
  secondaryInputBgdActive: '#F6F6F6',
  secondaryInputBgdHover: '#F6F6F6',
  secondaryInputBorderActiveColor: '#000000',
  secondaryInputBorderColor: 'none',
  secondaryInputColor: '#545454',
  sidePanelBg: '#F6F6F6',
  sidePanelHeaderBg: '#FFFFFF',
  subtextColor: '#AFAFAF',
  panelTabColor: '#AFAFAF',
  subtextColorActive: '#000000',
  textColor: '#000000',
  textColorHl: '#545454',
  mapPanelBackgroundColor: '#F6F6F6',
  mapPanelHeaderBackgroundColor: '#FFFFFF',
  titleTextColor: '#000000',
  switchBtnBgdActive: '#000000',
  switchBtnBgd: '#FFFFFF',
  switchTrackBgdActive: '#E2E2E2',
  secondarySwitchTrackBgd: '#E2E2E2',
  switchTrackBgd: '#E2E2E2',
  secondarySwitchBtnBgd: '#FFFFFF',
  histogramFillInRange: '#000000',
  histogramFillOutRange: '#E2E2E2',
  rangeBrushBgd: '#E2E2E2',
  sliderBarBgd: '#E2E2E2',
  sliderHandleColor: '#FFFFFF',
  sliderInactiveBorderColor: '#FFFFFF',
  sliderHandleTextColor: '#FFFFFF',
  sliderBarColor: '#000000'
});
