// kepler.gl/dist/dist/filters/range-filter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = RangeFilterFactory;
var _react = _interopRequireDefault(require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/node_modules/react"));
var _rangeSlider = _interopRequireDefault(require("../common/range-slider"));
var _constants = require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/src/constants/src");
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

RangeFilterFactory.deps = [_rangeSlider["default"]];
function RangeFilterFactory(RangeSlider) {
  var RangeFilter = function RangeFilter(_ref) {
    var filter = _ref.filter,
      setFilter = _ref.setFilter,
      setFilterPlot = _ref.setFilterPlot;
    return /*#__PURE__*/_react["default"].createElement("div", null, /*#__PURE__*/_react["default"].createElement(RangeSlider, {
      range: filter.domain,
      value0: filter.value[0],
      value1: filter.value[1],
      step: filter.step,
      bins: filter.bins,
      isEnlarged: filter.view === _constants.FILTER_VIEW_TYPES.enlarged,
      onChange: setFilter,
      setFilterPlot: setFilterPlot,
      inputTheme: "secondary",
      plotType: filter.plotType
    }));
  };
  return RangeFilter;
}
