// kepler.gl/dist/dist/filters/time-range-filter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
exports.timeRangeSliderFieldsSelector = timeRangeSliderFieldsSelector;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/node_modules/react"));
var _timeRangeSlider = _interopRequireDefault(require("../common/time-range-slider"));
var _constants = require("/Users/ihordykhta/Desktop/unfolded/kepler.gl/src/constants/src");
// SPDX-License-Identifier: MIT
// Copyright contributors to the kepler.gl project

/*
 * TimeRangeFilter -> TimeRangeSlider -> RangeSlider
 */
function timeRangeSliderFieldsSelector(filter, datasets, layers) {
  var hasUserFormat = typeof filter.timeFormat === 'string';
  var timeFormat = (hasUserFormat ? filter.timeFormat : filter.defaultTimeFormat) || _constants.DEFAULT_TIME_FORMAT;
  return {
    id: filter.id,
    domain: filter.domain,
    timeBins: filter.timeBins,
    value: filter.value,
    plotType: filter.plotType,
    lineChart: filter.lineChart,
    yAxis: filter.yAxis,
    step: filter.step,
    speed: filter.speed,
    animationWindow: filter.animationWindow,
    isAnimating: filter.isAnimating,
    timezone: filter.timezone,
    timeFormat: timeFormat,
    filter: filter,
    datasets: datasets,
    layers: layers,
    isMinified: filter.view === _constants.FILTER_VIEW_TYPES.minified,
    isEnlarged: filter.view === _constants.FILTER_VIEW_TYPES.enlarged
  };
}
TimeRangeFilterFactory.deps = [_timeRangeSlider["default"]];
function TimeRangeFilterFactory(TimeRangeSlider) {
  var TimeRangeFilterComponent = function TimeRangeFilterComponent(_ref) {
    var filter = _ref.filter,
      datasets = _ref.datasets,
      layers = _ref.layers,
      setFilter = _ref.setFilter,
      setFilterPlot = _ref.setFilterPlot,
      isAnimatable = _ref.isAnimatable,
      toggleAnimation = _ref.toggleAnimation,
      hideTimeTitle = _ref.hideTimeTitle,
      timeline = _ref.timeline;
    return /*#__PURE__*/_react["default"].createElement(TimeRangeSlider, (0, _extends2["default"])({}, timeRangeSliderFieldsSelector(filter, datasets, layers), {
      onChange: setFilter,
      setFilterPlot: setFilterPlot,
      toggleAnimation: toggleAnimation,
      isAnimatable: isAnimatable,
      hideTimeTitle: hideTimeTitle,
      timeline: timeline
    }));
  };
  return TimeRangeFilterComponent;
}
var _default = exports["default"] = TimeRangeFilterFactory;
