// @hubble.gl/core/dist/animations/animation-manager.js
// hubble.gl
// SPDX-License-Identifier: MIT
// Copyright (c) vis.gl contributors
import { Timeline } from '@luma.gl/engine';
export default class AnimationManager {
    constructor({ timeline = undefined, animations = [] }) {
        this.animations = {};
        this.timeline = timeline || new Timeline();
        for (const animation of animations) {
            this.attachAnimation(animation);
        }
    }
    attachAnimation(animation) {
        animation.attachKeyframes(this.timeline);
        this.animations[animation.id] = animation;
    }
    setKeyframes(animationId, params) {
        this.animations[animationId].setKeyframes({
            timeline: this.timeline,
            ...params
        });
    }
    getKeyframes(animationId) {
        return this.animations[animationId].getKeyframes();
    }
    getAnimation(animationId) {
        return this.animations[animationId];
    }
    draw() {
        Object.values(this.animations).forEach(animation => animation.draw());
    }
}
